{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "matplotlib.rcParams['xtick.labelsize'] = 9\n",
    "matplotlib.rcParams['ytick.labelsize'] = 9\n",
    "matplotlib.rcParams['axes.labelsize'] = 9\n",
    "matplotlib.rcParams['axes.titlesize'] = 10\n",
    "\n",
    "matplotlib.rcParams['axes.grid'] = True\n",
    "matplotlib.rcParams['grid.color'] = '0.5'\n",
    "matplotlib.rcParams['grid.linewidth'] = '0.5'\n",
    "\n",
    "matplotlib.rcParams['axes.edgecolor'] = '0.25'\n",
    "matplotlib.rcParams['xtick.color'] = '0'\n",
    "matplotlib.rcParams['ytick.color'] = '0'\n",
    "\n",
    "matplotlib.rcParams['xtick.major.width'] = 2\n",
    "matplotlib.rcParams['ytick.major.width'] = 2\n",
    "matplotlib.rcParams['ytick.major.size'] = 5\n",
    "matplotlib.rcParams['xtick.major.size'] = 5\n",
    "matplotlib.rcParams['axes.spines.right'] = True\n",
    "matplotlib.rcParams['axes.spines.left'] = True\n",
    "matplotlib.rcParams['axes.spines.top'] = True\n",
    "matplotlib.rcParams['axes.spines.bottom'] = True\n",
    "\n",
    "\n",
    "matplotlib.rcParams['font.family'] = 'sans-serif'\n",
    "matplotlib.rcParams['font.weight']='ultralight'\n",
    "# matplotlib.rcParams['font.sans-serif'] = ['Computer Modern Sans Serif']\n",
    "matplotlib.rcParams['axes.axisbelow'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import glob\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from scipy import stats\n",
    "import re\n",
    "import random\n",
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "from Bio.Data import CodonTable\n",
    "import itertools\n",
    "from collections import Counter\n",
    "\n",
    "from bisect import bisect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_codon_dicts(n=11):\n",
    "    GivenCodonTable = CodonTable.unambiguous_dna_by_id[n]\n",
    "    nucleotides = ['A', 'T', 'C', 'G']\n",
    "    aa_to_codons = {}\n",
    "    for first in nucleotides:\n",
    "        for second in nucleotides:\n",
    "            for third in nucleotides:\n",
    "                Codon = first + second + third\n",
    "                if Codon not in CodonTable.unambiguous_dna_by_id[n].stop_codons:\n",
    "                    if GivenCodonTable.forward_table[Codon] in aa_to_codons.keys():\n",
    "                            aa_to_codons[GivenCodonTable.forward_table[Codon]].append(Codon)\n",
    "                    else:\n",
    "                        aa_to_codons[GivenCodonTable.forward_table[Codon]] = [Codon]\n",
    "                else:\n",
    "                    pass\n",
    "    codon_to_aa = {}\n",
    "    for aa, syns in aa_to_codons.items():\n",
    "        for syn in syns:\n",
    "            codon_to_aa[syn] = aa\n",
    "    return aa_to_codons, codon_to_aa\n",
    "\n",
    "aa_to_codon_dict, codon_to_aa_dict = get_codon_dicts()\n",
    "\n",
    "temp = []\n",
    "for aa in ['P', 'T', 'A', 'V', 'G']:\n",
    "    temp.extend(aa_to_codon_dict[aa])\n",
    "\n",
    "redundant_codons = ['CTA', 'CTG', 'CTC', 'CTT',\\\n",
    "                    'TCA', 'TCG', 'TCC', 'TCT',\\\n",
    "                    'CGA', 'CGG', 'CGC', 'CGT'] + temp\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the sequences for each gene for testing purposes later\n",
    "Note: this requires establishing a reference genome that you will use later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of records: 61\n"
     ]
    }
   ],
   "source": [
    "base_genome = '511145.12' #This is E. coli K12\n",
    "species_name = 'Escherichia'\n",
    "seq_dicty = {}\n",
    "for infile_name in glob.glob('../Data/Orthologs/Order_Enterobacterales_Mixed/mfastas_nt/*.mfasta'):\n",
    "    gene_name = infile_name.split('/')[-1].strip('.mfasta')\n",
    "    records = list(SeqIO.parse(infile_name, 'fasta'))\n",
    "    for record in records:\n",
    "        try:\n",
    "            seq_dicty[record.id][gene_name] = str(record.seq)\n",
    "        except KeyError:\n",
    "            seq_dicty[record.id] = {}\n",
    "            seq_dicty[record.id][gene_name] = str(record.seq)\n",
    "print('Number of records: {}'.format(len(seq_dicty.keys())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#######To analyze a different genome as a reference, run this cell instead and \n",
    "#######just pretend that it's E. coli which makes the code easier\n",
    "# base_genome = '511145.12' #This is E. coli K12\n",
    "# new_base_genome = '214092.21'\n",
    "# species_name = 'Yersinia'\n",
    "# seq_dicty = {}\n",
    "# # for infile_name in glob.glob('../Data/Orthologs/Order_Enterobacterales_Reference/mfastas_nt/*.mfasta'):\n",
    "# # for infile_name in glob.glob('../Data/Orthologs/Order_Enterobacterales_Representative/mfastas_nt/*.mfasta'):\n",
    "# for infile_name in glob.glob('../Data/Orthologs/Order_Enterobacterales_Mixed/mfastas_nt/*.mfasta'):\n",
    "# #     print(infile_name)\n",
    "#     gene_name = infile_name.split('/')[-1].strip('.mfasta')\n",
    "#     records = list(SeqIO.parse(infile_name, 'fasta'))\n",
    "#     for record in records:\n",
    "#         if record.id not in [base_genome, new_base_genome]:\n",
    "#             try:\n",
    "#                 seq_dicty[record.id][gene_name] = str(record.seq)\n",
    "#             except KeyError:\n",
    "#                 seq_dicty[record.id] = {}\n",
    "#                 seq_dicty[record.id][gene_name] = str(record.seq)\n",
    "#         elif record.id == base_genome:\n",
    "#             try:\n",
    "#                 seq_dicty[new_base_genome][gene_name] = str(record.seq)\n",
    "#             except KeyError:\n",
    "#                 seq_dicty[new_base_genome] = {}\n",
    "#                 seq_dicty[new_base_genome][gene_name] = str(record.seq)\n",
    "#         elif record.id == new_base_genome:\n",
    "#             try:\n",
    "#                 seq_dicty[base_genome][gene_name] = str(record.seq)\n",
    "#             except KeyError:\n",
    "#                 seq_dicty[base_genome] = {}\n",
    "#                 seq_dicty[base_genome][gene_name] = str(record.seq)\n",
    "                \n",
    "# print(len(seq_dicty.keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## And get dicts to identify conserved amino acid positions and consensus nucleotides\n",
    "\n",
    "Please note that I'm considering an amino acid here as being \"conserved\" if 60 positions in the alignment have that same amino acid (and it's not a gap). There are 61 total positions so this just gives a little bit of flexibility to definition of conserved."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "conservation_dicty = {}\n",
    "nt_consensus_dicty = {}\n",
    "nt_conservation_dicty = {}\n",
    "for infile_name in glob.glob('../Data/Orthologs/Order_Enterobacterales_Mixed/mfastas_nt/*.mfasta')[:]:\n",
    "    gene_name = infile_name.split('/')[-1].strip('.mfasta')\n",
    "    conservation_dicty[gene_name] = []\n",
    "    nt_consensus_dicty[gene_name] = []\n",
    "    nt_conservation_dicty[gene_name] = []\n",
    "    records = list(SeqIO.parse(infile_name, 'fasta'))\n",
    "    aa_seq_lol = []\n",
    "    nt_seq_lol = []\n",
    "    ref_genome_aa = []\n",
    "    ref_genome_nt = []\n",
    "    for record in records:\n",
    "        aa_seq_lol.append(list(str(record.seq.translate(gap='-'))))\n",
    "        nt_seq_lol.append(list(str(record.seq)))\n",
    "        if record.id == base_genome:\n",
    "            ref_genome_aa = list(str(record.seq.translate(gap='-')))\n",
    "            ref_genome_nt = list(str(record.seq))\n",
    "    aa_seq_lol = np.array(aa_seq_lol)\n",
    "    aa_seq_lol = aa_seq_lol.T\n",
    "    for i, pos in enumerate(aa_seq_lol):\n",
    "        counter_dict = Counter(pos)\n",
    "        best = sorted(counter_dict.items(), key=lambda x: x[1])[-1]\n",
    "        if len(counter_dict.keys()) == 1 and counter_dict[list(counter_dict.keys())[0]] != '-':\n",
    "            conservation_dicty[gene_name].append('C')\n",
    "        elif len(counter_dict.keys()) == 2 and best[1]==60 and\\\n",
    "                best[0]!='-' and best[0]==ref_genome_aa[i]:\n",
    "            conservation_dicty[gene_name].append('C')\n",
    "        else:\n",
    "            conservation_dicty[gene_name].append('V')\n",
    "    nt_seq_lol = np.array(nt_seq_lol)\n",
    "    nt_seq_lol = nt_seq_lol.T\n",
    "    for i,pos in enumerate(nt_seq_lol):\n",
    "        counter_dict = Counter(pos)\n",
    "        nt_consensus_dicty[gene_name].append(sorted(counter_dict.items(), key=lambda x: x[1])[-1][0])\n",
    "        best = sorted(counter_dict.items(), key=lambda x: x[1])[-1]\n",
    "        if len(counter_dict.keys()) == 1 and counter_dict[list(counter_dict.keys())[0]] != '-':\n",
    "            nt_conservation_dicty[gene_name].append('C')\n",
    "        elif len(counter_dict.keys()) == 2 and best[1]==60 and\\\n",
    "                best[0]!='-' and best[0]==ref_genome_nt[i]:\n",
    "            nt_conservation_dicty[gene_name].append('C')\n",
    "        else:\n",
    "            nt_conservation_dicty[gene_name].append('V')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read in substitution rate datasets as separate dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# dicty_phylop = {}\n",
    "# for infile_name in glob.glob('../Results/Order_Enterobacterales_Mixed/phyloP_CONACC/*.wig')[:]:\n",
    "#     with open(infile_name, 'r') as infile:\n",
    "#         tempy = infile.readlines()\n",
    "#         data = [float(i) for i in tempy[1:]]\n",
    "#         dicty_phylop[infile_name.split('/')[-1].strip('.wig')] = data\n",
    "# print(len(dicty_phylop.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1394\n"
     ]
    }
   ],
   "source": [
    "dicty_hyphy = {}\n",
    "for infile_name in glob.glob('../Results/Order_Enterobacterales_Mixed/Hyphy_GTR/*.site-rates.json')[:]:\n",
    "    gene_name = infile_name.split('/')[-1].strip('.site-rates.json')\n",
    "    with open(infile_name) as infile:\n",
    "        site_rate_data = json.load(infile)\n",
    "    positions = [int(i) for i in site_rate_data['Relative site rate estimates'].keys()]\n",
    "    rates = []\n",
    "    for position in range(min(positions), max(positions)+1):\n",
    "        rates.append(site_rate_data['Relative site rate estimates'][str(position)]['MLE'])\n",
    "    meany = np.mean(rates)\n",
    "    dicty_hyphy[gene_name] = [i/meany for i in rates]\n",
    "print(len(dicty_hyphy.keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Trim 5' terminal gaps according to the E. coli reference sequence for all dicts instantiated above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1105.7195122\n",
      "1105.7195122\n",
      "1059.24533716\n",
      "1059.24533716\n"
     ]
    }
   ],
   "source": [
    "print(np.mean([len(i) for i in seq_dicty[base_genome].values()]))\n",
    "print(np.mean([len(i) for i in dicty_hyphy.values()]))\n",
    "\n",
    "for i, j in seq_dicty[base_genome].items():\n",
    "    if j[0:3] == '---':\n",
    "        first_bases = []\n",
    "        for base in ['A', 'T', 'G', 'C']:\n",
    "            first_bases.append(j.find(base))\n",
    "        first_base = min(first_bases)\n",
    "        \n",
    "        for genome in seq_dicty.keys():\n",
    "            seq_dicty[genome][i] = seq_dicty[genome][i][first_base:]\n",
    "        \n",
    "        if i in dicty_hyphy.keys():\n",
    "            dicty_hyphy[i] = dicty_hyphy[i][first_base:]\n",
    "\n",
    "        if i in conservation_dicty.keys():\n",
    "            conservation_dicty[i] = conservation_dicty[i][int(first_base/3):]\n",
    "                \n",
    "        if i in nt_consensus_dicty.keys():\n",
    "            nt_consensus_dicty[i] = nt_consensus_dicty[i][first_base:]\n",
    "            \n",
    "        if i in nt_conservation_dicty.keys():\n",
    "            nt_conservation_dicty[i] = nt_conservation_dicty[i][first_base:]\n",
    "            \n",
    "print(np.mean([len(i) for i in seq_dicty[base_genome].values()]))\n",
    "print(np.mean([len(i) for i in dicty_hyphy.values()]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## And 3' terminal gaps according to the E. coli reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1059.24533716\n",
      "1059.24533716\n",
      "1046.48995696\n",
      "1046.48995696\n"
     ]
    }
   ],
   "source": [
    "print(np.mean([len(i) for i in seq_dicty[base_genome].values()]))\n",
    "print(np.mean([len(i) for i in dicty_hyphy.values()]))\n",
    "\n",
    "\n",
    "for i, j in seq_dicty[base_genome].items():\n",
    "    if j[-3:] == '---':\n",
    "        last_bases = []\n",
    "        for base in ['A', 'T', 'G', 'C']:\n",
    "            last_bases.append(j[::-1].find(base))\n",
    "        last_base = min(last_bases)\n",
    "        for genome in seq_dicty.keys():\n",
    "            seq_dicty[genome][i] = seq_dicty[genome][i][:len(j)-last_base]\n",
    "\n",
    "        if i in dicty_hyphy.keys():\n",
    "            dicty_hyphy[i] = dicty_hyphy[i][:len(j)-last_base]\n",
    "\n",
    "        if i in conservation_dicty.keys():\n",
    "            conservation_dicty[i] = conservation_dicty[i][:int((len(j)-last_base)/3)]\n",
    "\n",
    "        if i in nt_consensus_dicty.keys():\n",
    "            nt_consensus_dicty[i] = nt_consensus_dicty[i][:len(j)-last_base]\n",
    "        \n",
    "        if i in nt_conservation_dicty.keys():\n",
    "            nt_conservation_dicty[i] = nt_conservation_dicty[i][:len(j)-last_base]\n",
    "            \n",
    "print(np.mean([len(i) for i in seq_dicty[base_genome].values()]))\n",
    "print(np.mean([len(i) for i in dicty_hyphy.values()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Investigate relationships between the metrics on the full dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x120423860>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAELCAYAAADDZxFQAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGVVJREFUeJzt3X2QHVd55/Hfz0YW2Ai9hSS8RDYuzFtwYmRHlTJ+uWze\njHdlXLVexhvLtsqKPC42COdFm9Sugq1d2AKykKzZqBLLVrxgkFmIV8FQROBEd6xIRsRgJ4F4bZxo\nYEUIJbDkWIyxbPLsH30ubkY9M7fvTJ87LX0/Vbfm9rndp5/pe+c+c87pPu2IEAAAk5007AAAAPMT\nCQIAUIkEAQCoRIIAAFQiQQAAKpEgAACVSBAAgEokCABAJRIEAKASCQIAUIkEAQCo9LxhB1CX7dWS\nVl9yySXrV6xYMVAdExMTOvXUU+c2sDlAXPUQVz3EVc/xGtejjz767V27dr24r5UjopWP66+/Pga1\na9eugbdtEnHVQ1z1EFc9x2tcnU7ngejze5YuJgBAJRIEAKASCQIAUKl1CcL2atu3PvHEE8MOBQCO\na61LEBFxT0Rcv3jx4mGHAgDHtdYlCABAHiQIAEAlEgQAoFLrrqSeC2NjYxobGxto25tuummOowGA\n+YkWBACg0tBbELZPl7RO0hOStkfEPw45JACAGmxB2N5p+6DtTaWytbb32t5je2UqfpukQymWp5uK\nBwBQT5NdTOskbewt2F4qaYOkjqQ1km5JL50i6T5Jfyrp2qqKbMfkR4NxAwDUYBdTRBywXS5aJWl3\nRByVtN/2ItsLJf2xpKslPStpe7/1T0xMqNvtzmHE/Wlyn+Pj40P5nWZCXPUQVz3EVU/OuHKOQSxX\n0ZXUc1jSsoj4G5VaGlUi4geZpnc/iGeeeWZ9p9MZKJBBz2CSpEH32Y9ut9to/YMirnqIqx7iqme2\ncW3evLnvdXOexfS4pCWl5cWprBam2gCAPHImiH2SLrC9wPYKSUciovagNJP1AUAeTZ7FtFVF19Fa\n2zsi4pCkLZLGVIw13DhIvbQgACCPJgep11eUbZO0bTb19sYgRkZGZlMNAGAGrbuSmhYEAOTRugTB\nGAQA5NG6BEELAgDyaF2CAADk0boEQRcTAOTRugRBFxMA5NG6BAEAyKN1CYIuJgDIo3UJgi4mAMij\ndQkCAJAHCQIAUKl1CYIxCADIo3UJgjEIAMijdQkCAJAHCQIAUIkEAQCo1LoEwSA1AOTRugTBIDUA\n5NG6BAEAyIMEAQCoRIIAAFQiQQAAKpEgAACVnjfsAGyvlfSTkr4h6X9ERAw3IgCA1GALwvZO2wdt\nbyqVrbW91/Ye2ytT8T9L+q6khZJObioeAEA9TXYxrZO0sbdge6mkDZI6ktZIukWSIuLuiLhZ0mOS\nfqGqItsx+dFg3AAANdjFFBEHbJeLVknaHRFHJe23vcj2QklvknSOpDMlbTq2pmoTExPqdrtzGHF/\nmtzn+Pj4UH6nmRBXPcRVD3HVkzOunGMQyyUdKi0flrQsIv5M0p9Nt2FEeHLZ6OhodDqdgQIZGxsb\naDtJGnSf/eh2u43WPyjiqoe46iGuemYb1+bNm/teN+dZTI9LWlJaXpzKamEuJgDII2eC2CfpAtsL\nbK+QdCQins64fwBADU2exbRVxSD1Wts7IuKQpC2SxiRtl3TjIPUyWR8A5NHkIPX6irJtkrbNpl7b\nqyWtHhkZmU01AIAZtO5KaloQAJBH6xIEg9QAkEfrEgQtCADIo3UJAgCQR+sSBF1MAJBH6xIEXUwA\nkEfrEgQAII/WJQi6mAAgj9YlCLqYACCP1iUIAEAeJAgAQKXWJQjGIAAgj9YlCMYgACCP1iUIAEAe\nJAgAQCUSBACgUusSBIPUAJBH6xIEg9QAkEfrEgQAIA8SBACgEgkCAFCJBAEAqESCAABUmhcJwvb5\ntr847DgAAM9pLEHY3mn7oO1NpbK1tvfa3mN7ZSo7S9JrJf1VU7EAAOprsgWxTtLG3oLtpZI2SOpI\nWiPplvTSZZIWSTrb9i9WVWQ7Jj8ajBsAIOl5TVUcEQdsl4tWSdodEUcl7be9yPbCiHi/JNl+TUR8\ntt/6JyYm1O125zTmfjS5z/Hx8aH8TjMhrnqIqx7iqidnXI0liArLJR0qLR+WtEzSNyUpIm6YasOI\n8OSy0dHR6HQ6AwUyNjY20HaSNOg++9Htdhutf1DEVQ9x1UNc9cw2rs2bN/e9bs5B6sclLSktL05l\ntTAXEwDkkTNB7JN0ge0FtldIOhIRT2fcPwCghibPYtqqYpB6re0dEXFI0hZJY5K2S7pxkHqZrA8A\n8mhykHp9Rdk2SdtmU6/t1ZJWj4yMzKYaAMAM5sWFcnXQggCAPFqXIBikBoA8WpcgaEEAQB6tSxAA\ngDxalyDoYgKAPFqXIOhiAoA8WpcgAAB5tC5B0MUEAHm0LkHQxQQAebQuQQAA8iBBAAAq9ZUgbL89\n/Xyj7QdsX9tsWNPGwhgEAGTQbwvi8vRzVNJbJE15c5+mMQYBAHn0myBOs32apGcj4huSuI8DABzn\n+k0Qd0naJek22y+Q9O3mQgIAzAf93g/ikYhY1Vuw/QcNxQMAmCf6bUFsnLT8trkOBAAwv0ybIGxf\nbftzks6x/dn02CXptDzhVcbEWUwAkMFMLYhPSvoVSR+WtD49flnSv244rilxFhMA5DFtgoiIJyLi\naxHxDhWthhWSzpJ0YY7gAADD09cgte2PSlou6eupKCTd11RQAIDh6/csppdGRKfJQAAA80u/ZzHd\nb/vsRiMBAMwr/bYgrpB0pe1n03JExKvmIgDbr5Z0qYrxjVsj4uG5qBcAMDt9tSAi4qyIeEX6eVY/\nycH2TtsHbW8qla21vdf2HtsrU92PSPqaigTxzIC/BwBgjvU7SH3N5LKI+NAMm62T9POSXp7qWCpp\ng6SflfQyFafOXpDqutv230t6g6THKvYfk8uuv/76fkIHAAyo3y6mn0g/LelcSd+TNG2CiIgDtstF\nqyTtjoijkvbbXmR7oYpTZldKOl3Srf0GPjExoW632+/qc6bJfY6Pjw/ld5oJcdVDXPUQVz054+or\nQUTEu8vLtj8xwL6WSzpUWj4saVlE3Cvp3hn278llo6Oj0el0BghDGhsbG2g7SRp0n/3odruN1j8o\n4qqHuOohrnpmG9fmzZv7XrffGwa9tPRYJenVA8T1uKQlpeXFqawWptoAgDz67WL6SPoZKr7U/8MA\n+9on6V22F0h6iaQjEcF9JQBgnuq3i+lNdSu2vVXS+ZIW2j4vIi63vUXSmIpE8466daZY7pF0z+jo\n6PpBtgcA9KffLqbzbO+z/Uj6uWqmbSJifUT8ZES8MiIuT2XbIuL8iHhjRDwwSMB0MQFAHv1eSf37\nkq6MiFdL+veSfq+5kKbHbK4AkEe/CSIiYn968g+S/qW5kKZHCwIA8ug3QXzF9m22r7N9u6S/azKo\n6dCCAIA8+j2L6W5JL5T0SkmfknSksYgAAPNCvwnityLi53oLtu+V9LlmQpqe7dWSVo+MjAxj9wBw\nwui3i+kFk5ZPnetA+kUXEwDk0W8L4j7bd0raJeliSbubCwkAMB/0O933b6u4mnqZpO0R8VuNRjUN\nzmICgDz67WJSRHwmIn43Ij7TZEB9xEEXEwBk0HeCAACcWEgQAIBKrUsQjEEAQB6tSxCMQQBAHq1L\nEACAPEgQAIBKJAgAQCUSBACgUusSBGcxAUAerUsQnMUEAHn0O1kfks2bN89q+5tuummOIgGAZrWu\nBQEAyIMEAQCoRIIAAFQa+hiE7Qslna/iftfvjYjHhhwSAEANtiBs77R90PamUtla23tt77G9UpIi\nYndEvFfSg5Je2lQ8AIB6muxiWidpY2/B9lJJGyR1JK2RdEvptX8r6aSIuK+qItsx+dFg3AAANdjF\nFBEHbJeLVknaHRFHJe23vcj2QkmrJf2qpB22z46Iv+2n/omJCXW73bkOu3HTxTw+Pj4vfyfiqoe4\n6iGuenLGlXMMYrmkQ6Xlw5KWRcQnJH1iug0jwpPLRkdHo9PpDBTI2NjYQNvNheli7na7074+LMRV\nD3HVQ1z1zDauOtdy5TyL6XFJS0rLi1NZLUy1AQB55EwQ+yRdYHuB7RWSjkTE0xn3DwCoocmzmLaq\nGKRea3tHRByStEXSmKTtkm4cpF7mYgKAPJocpF5fUbZN0rbZ1Gt7taTVIyMjs6kGADCDoV8oV1dE\n3CPpntHR0WMSUBvMNEA03QA6E/0ByKl1U20wSA0AebQuQTAGAQB5tC5BAADyaN0YxIk8SD2bmxUx\nfgGgrta1IOhiAoA8WpcgAAB5tC5BcBYTAOTRujGItl8HMSyzGb+4+OKL5zASAG3RuhYEACAPEgQA\noFLrEgRjEACQR+sSBKe5AkAerUsQAIA8SBAAgEqtO80V+Y2NjQ18H2+m+ADaixYEAKBS6xIEZzEB\nQB6t62LiSup2mc0V3BJdVMAwta4FAQDIgwQBAKhEggAAVCJBAAAqDT1B2H697TtsXznsWAAAz2ks\nQdjeafug7U2lsrW299reY3ulJEXElyXd0VQcAIDBNNmCWCdpY2/B9lJJGyR1JK2RdEu/FdmOyY+5\nDhYA8MMauw4iIg7YLhetkrQ7Io5K2m97ke2Fkl4s6QpJi2x/KSIe7af+iYkJdbvduQ4b88x07/H4\n+Pi8/AwQVz3EVU/OuHJeKLdc0qHS8mFJyyLigKRfnW7DiPDkstHR0eh0OgMFMui8Qshvuve42+1O\n+/qwEFc9xFXPbOOqc/FqzgTxuKQlpeXFqawW26slrR4ZGZmruHCcms1V3FzBDeRNEPskvcv2Akkv\nkXQkIp7OuH+00Exf8rQGgeY0eRbTVhWD1Gtt74iIQ5K2SBqTtF3SjYPUyx3lACCPJgepj5lMLyK2\nSdo2m3rpYgKAPIZ+oVxdtCAAII/WJQjuBwEAebQuQdCCAIA8WpcgAAB5tC5B0MUEAHm0LkHQxQQA\nebQuQQAA8sh5JfWc4DoI5DCbaTqk4V3hzRQhmEuta0HQxQQAebQuQQAA8iBBAAAqMQYBHEdmM/vt\nbMYvZjNmc/HFFw+8LZrVuhYEYxAAkEfrEgQAIA8SBACgEgkCAFCJBAEAqMRZTACQWVvO+mpdC4Kz\nmAAgj9YlCABAHiQIAEAlEgQAoBIJAgBQaehnMdk+R9KIpIWS3h8R3xhySAAANdiCsL3T9kHbm0pl\na23vtb3H9spUfK2kmyX9kaQrmooHAFBPk11M6yRt7C3YXippg6SOpDWSbkkvxUwV2Y7JjwbiBQCU\nNNbFFBEHbJeLVknaHRFHJe23vcj2QkkfknSTii6mD/Rb/8TEhLrd7hxGDJzYZnub1UGNjY3N6hat\nTV04Nj4+Pi+/Y3LGlXMMYrmkQ6Xlw5KWRcRDkh6absOI8OSy0dHR6HQ6AwUyrPsFA5h7g34PzKTb\n7TZW92y+g84444xZxVXnH4GcZzE9LmlJaXlxKqvF9mrbtz7xxBNzFhgA4Fg5E8Q+SRfYXmB7haQj\nEfF03UqYagMA8misi8n2VknnS1po+7yIuNz2FkljKgam3zFgvUzWBwAZNDlIvb6ibJukbU3tEwAw\nd1p3JTVdTACQR+sSBAAgj9YlCM5iAoA8Wpcg6GICgDwc0a5ZK3pnMV100UUjJ5100lfrbv/kk0+e\nK0mLFi364pwHNwvEVQ9x1UNc9RzncZ2+a9euF/ezYusSxGz15nGqujp7mIirHuKqh7jqIa5C67qY\nAAB5kCAAAJVIEACASifcGAQAoD+0IAAAlUgQAIBKJAgAQKUTLkHYXmt7r+09tldm3vcb0n7vs/0X\nts9M8ey33U2Pl6V1z0jr7LH9nzLE9lQphnUufND2btufsr0srbcsLe9Orzd2Prbt15Viut/2d2x3\nbH+zVH5urrhs77R90PamtFzrGNlemd7PvbbXNhjXNba/kD5nd6Vb+8r2HbYfTMft46XtL0nH937b\nv9RgXLXeu4zH65dLMT1s+09S+c1puffayU3FNcV3w/NtfyQdl4/Yfn5at/K7oZH3MSJOmIekpZK+\nJOkUSa+Q9JeZ9//jkhal55dK+rCktZI2Vax7l6QL0/N7Jb2m4dgem7R8iaTb0/NrJL0nPX+PpKvT\n822SLsl07N4q6Q8ldSTdVvF643FJenn5/ap7jCTtSZ+7U9LncGlDcZ0p6eT0/H2S1qXnd0i6YNK2\nJ0v6GxV3e1wi6a972zYQV633LtfxmvTaFklXpuc3S1pTsc6cxzXFd8MNkn4nlb1T0g3p+THfDU29\njydaC2KVpN0RcTQi9kta1PvvKoeI+KeIeDItPi3p2fT8Gtt/afu/2u69J+dExO70/NOSmrkz+3N+\n3PaY7bttn5H296n02j2l/U9V3rQ1ku5Mz3+p9N/mC3LFFREHJhX1fYzS5+y0iNgfEUcl7VbxeZzz\nuCLiHyLi+2mx/DmTpA+kY9e749YrJe2PiMMRcVjSeCqb87iSvt67nMerx/YCSW+W9Kel4v+Y/jY3\npHUaiWuK74apPl9V3w2NvI8nWoJYLulQafmwpGW5g7B9mqR3SfpdFR/G16p4k0+XdFVarfze5Ijz\njIi4WNIfSbpdP3ysDqtofSnFcThjXLK9XMV/SXskfVHSWRFxoaR/lvSbw4pL9Y7R8lJZlhhtv0ZF\nK+djqeg3I2KVpLdI+m3bZyrv30Sd9y778VKRHO6LiKfS8gcl/bSkX5B0me0Lm45r0nfD5M9Xbz9V\n3w2NvI8nWoJ4XEXzq2dxKssm/ZfyMUnvjYi/i4hDEfH99B/fXZLOS6v+S844I+Lb6edOFYmqfKwW\n67kP36G0nCWuZETSx6PwZER8L5V/RM8dr2HEVecYZf3s2X65pP+lorvke9IPvcePS/qcii+/bHHV\nfO+G8bdabqUqIr6TPnNPSbo7xdtYXJO/G3Ts56u3n6rvhkbiOtESxD5JF9heYHuFpCMR8XSunafu\nozsl7YiIHams/Kb+K0mPpOd/bfv89PzNku5rMK4XlgbgfkrSt1XcO/zStMqlaVnTlDfpKqU/XNvl\ned7Lx2sYcfV9jNIX43dtr0hfBBdI+kITQdn+EUl/oqLP+u9L5UvSz1MkvVHSo5K+KukVtl9k+0Uq\n+tYfayiuvt+7nMcrxfYiSedK+vNSWe94WcX4ySNNxVX13aCpP19V3w3NvI9zMejTpoek6yTtVdFd\ncV7mfV8h6Yikbnp8UNK7VSSuPSoG6Bakdc+UtCuVHzOYNsdxrZL0YPqgjan4z/IkSX+goo/105KW\np3WXp+Xd6fWTGo7tTEkPlJbfJumBFOvdkpbkikvSVklfSX94O+oeIxX/ge5Jn7/rGozrf0o6UPqc\n9Qapd6b9f0HShtL2l0q6Pz0ubTCuWu9druOVyq6T9P5J692RjsnnlU5AaCouVX83vEDS9nRctkt6\nfulv4pjvhibeR6baAABUOtG6mAAAfSJBAAAqkSAAAJVIEACASiQIAEAlEgTmpTQh2b2zrKNj+7YB\ntx2xfV2N9df2Jn8bcH83214z6PYz1L2hj3W2pYvrgB8gQQDVblAxYdoPKc2VNa/0LnScwowJQsW1\nAb8xR+HgODEvP+xAstT2nba/ZPtGSUoTvf1oen6h7dvT86/b/uM01fH7SnW8wvb/tv23tv+d7ZNt\nP5SuJpbtq23/Tnmntl8r6VsR8Uxa7tp+v+2dkl5p++0pjvtt/8pUwds+1fZn0iSIXduvKrc0bL/c\ndre0yWW2P217n4upzm37o2lfu2xfNKn+joupqz8u6d0upq3eleK6LW3/65Je5uemcV+cjsefu5gy\nujeh2+cl/VzN9wfHuecNOwBgGi+RdKGKuWcelvT7Kq5uvUbSf5e0TsUU4L11b5L0/yTttH1OKl8i\n6Rcl/ZikT0bEx23fI+kySZ+QdG16lL1ex05T8EBE/EZKHpdIukjFP1i7bf+fKeJ/jaRDEfFm6Qet\nj/OnWFdp3bfafqOk/5Z+v9NVTNEdU7ReXirp30TEM7ZPi4iPpn19TMWU0B+w/baI6KTy90i6OyLu\nsv3TKqbaviLV/5TtpRFxqGI/OAGRIDCfPRwRE5Jkuzd99V2S/sL2rSrukfH5VP5PEfH1tO4XJL1a\n0rckPRTFRIj/6OfmvbpN0hbbD0qaiIhv9BHL3vTz9ZJep2KqA0l6kaSfmGKbByV90fadkr6jIoGV\npy6YfFOj3pw++yS9KiK+Y3urpA/bnpD0X1RMoVH2QK+lI+ki2xtV3BvgdEmfrIjpbBXTad+Qlp+t\nWAeQRBcT5rdj5oGJiO+quEnLLSrmp+n5sdIg63kqJi+bqo6vpfKbVExtPtmXdexc+r0E9bCKL/43\npf/K3xARD00R/0JJH4iINZIOSrpaxQybvTjPnbR+b3bTn5H01TQZ3J1p+/sk/VrFPr5fev4eSVdF\nMW37Pj2XgMqzf35F0vsiopPiv1T6wYR0p9J6QBktCLTRrSomJPv1Utk3Jb3T9tmS9kbEl2x3pqnj\ndhUTwx1zplJEPGz7JbZPieKmMOXXvpzOrhpLrZqnbF82xT5eJ+kW28+q+Gfs2hTnr9n+rIpEU/ZC\n25+R9CMq7nj2o5LuSvs5RTMPNn9I0uds/99J5fenbrCPqZgc8g9tv11FAvm0iu66n1VpJlNAEpP1\noX3S+MLGiLiqVPZYRPR9By3bl0v6mYj4z1O8fqWKO4dVtTCOO7a3SXpnTHG3NZyYSBBoFdtXSbpR\n0rVR3FSlV953gkhn9rxV0lsi4lvNRAq0HwkCAFCJQWoAQCUSBACgEgkCAFCJBAEAqESCAABUIkEA\nACr9fz97o7cMNJptAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1099dd550>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hyphy = []\n",
    "for key, val in dicty_hyphy.items():\n",
    "    hyphy.extend(dicty_hyphy[key])   \n",
    "fig, ax = plt.subplots(figsize=(6,4))\n",
    "ax.hist(hyphy, 20, color='gray', alpha=1)\n",
    "ax.set_xlabel('hyphy (rel subs rate)')\n",
    "ax.set_yscale('log')\n",
    "ax.set_ylabel('counts')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Look at length and position-dependent scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ivals = []\n",
    "plot_vals_hyphy = []\n",
    "\n",
    "for i in range(150):\n",
    "    tempy = []\n",
    "    for gene_name, vals in dicty_hyphy.items():\n",
    "        if len(vals) > 150:\n",
    "            tempy.extend([vals[i]])\n",
    "    plot_vals_hyphy.append(np.median(tempy))\n",
    "    \n",
    "    ivals.append(i)\n",
    "    \n",
    "ivals_r = []\n",
    "plot_vals_hyphy_r = []\n",
    "for i in range(150, 0, -1):\n",
    "    tempy = []\n",
    "    for gene_name, vals in dicty_hyphy.items():\n",
    "        if len(vals) > 150:\n",
    "            tempy.extend([vals[-i]])\n",
    "    plot_vals_hyphy_r.append(np.median(tempy))\n",
    "    ivals_r.append(-i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZYAAAC6CAYAAACN4WtvAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJztnXmYFdWZ8H9vN70ioN0I2Eg3yKbGBW0hSlxwyWicjDGf\nkjGhiSYaJHEMwWViJB9GjMvEXScOUfQToeMaJwvRiSINIigIKk4cbRDtJjqsDS6xF6B5vz+qCm5f\nqu6t233X7vf3POepqlNneavqVL11tveIqmIYhmEYySIv0wIYhmEY3QtTLIZhGEZSMcViGIZhJBVT\nLIZhGEZSMcViGIZhJBVTLIZhGEZSMcViGIZhJBVTLIZhGEZSMcViGIZhJBVTLIZhGEZSMcViGIZh\nJJVemRYgG3jjjTfOz8/PvxPIz7QshmEYIWhvb2+/+vjjj/99pgXxwxQL0KtXr+uGDx/e0rt379ZM\ny2IYhhGPL774onj9+vXXAVmpWKwpDFDVQaWlpaZUDMPICUpLS1tVdWCm5QjCFIuLiGRaBMMwjFC4\n36us/WiZYkkTDQ0NBUceeeQRRUVFx+/atcs3zLZt2/Lnzp17YLLyXL58ecndd9/dH+Cee+4p9/wv\nvvjiIcnKo6ewaNGi3scdd9zh1dXVoy+99NJQ92/BggV9rrrqqopk5G/PMjm8/vrrxd5zvPDCC4fu\n2bMnZvhx48aNTlbe3rO67777ytvb2wG4/vrrB3344YcFycojWzDFEpLZsymrqODovDyqKyo4evZs\nyhKJP2DAgN1LlixZe+yxx34RFKapqSn/2WefPajr0jqMHz++Zfr06dsA5s2b19/znzt37t+SlUcu\nMPv12WUVd1YcnXdjXnXFnRVHz359dkLPDmDEiBFty5Ytq1+9enX91q1be61cubIkFbIGYc/SZfbs\nMioqjiYvr5qKiqOZndizPOaYY9refPPN91avXl0PsHTp0tLI894HPxV4z2r+/Pn9PYV2yy23bBo2\nbJj/n2YOY533IZg9m7Lp06lqbXUU8caNFE6fThXA1KlsD5NGaWmplpaWdii1L774Yu+rr756SElJ\nyZ6LLrpo+/r164teeeWVvuPGjRv9+9//fn1FRcVugKuuuqpi3bp1RU1NTQWHHnpo21NPPdXY1NSU\nf+GFFw77/PPP84866qjmRx999G+PPfbYgbfffvshvXv3bp8+ffrm3r1773nhhRf6VFZW7ly7dm3J\nuHHjRv/7v//7hh/+8IeVq1evrn/11VdLrrjiiqr29nYuv/zyLT/60Y+2X3DBBUOLior21NfXl0yY\nMOGzO++8c2Oy72c6mf367LLpL0yvat3d6jy7v28snP7CdOfZjZ0a6tkBVFZW7vb2CwoKND8/X+vr\n6wu/853vDOvfv//ujz76qPCPf/zj+8OHD981ceLEoR9//HHhwQcfvGv48OFtkenYs+wCs2eXMX16\nFa3Os2TjxkKmO8+SqeGeZVFR0d4lcwsLC/cMGzZs54IFC/rcddddA51kpm7ZsGFD4bx58/qfcMIJ\n+/0ELliwoM+tt946KD8/n507d8of/vCH9QMHDmy/5JJLhvz1r38t7dOnT/szzzzz4caNG3vV1NQM\nKyoq2nP66ad/dtttt22qrq4efccdd/zt3XffLRk/fvzoSy65ZOuiRYv63njjjRsPOeSQ3dFl4L77\n7iv/85//fGBra6sA1NXVvZ+Xlxt1gbhSishhIvK0iPxFRHqJyFXpECybmDWLwZ5S8WhtJW/WLAZ3\nJd0FCxb0u+WWWz5esWLF2mnTpm278sort5588smfrVy5st5TKh5HHnlky/Lly9cWFBToSy+91Pue\ne+7pf+GFF25ftWpVfUtLS96iRYt6P/vsswc988wz61977bW1EydO/NSLe80112wbNWpUy8qVK+vH\njRvX4vnPmDFj8OOPP/7BypUr62fPnj2gra1NAM4555zPVq9eXf/iiy/268r1ZQOzXp412FMqHq27\nW/NmvTyrU89uxYoVJU1NTb2qq6tbAZqbm/Ofe+659f/yL/+y+be//e1BdXV1pfn5+bp8+fK1Rxxx\nRItfGvYsO8msWYP3KhWP1tY8ZiX2LGtra/uNHDnyS1u3bi0YOHBgO8CuXbtk0aJF73/zm9/87LHH\nHuu/atWq9y666CJfZaWqvPzyy+suvfTSrffee+/BS5YsKW1ubs5btWpV/cSJE7ffddddB7/44ot9\nvv/9729dsWLF2ltuuWWTF/f0009vPuKII1qWL19ef8UVV+xN368MAFRUVOxcsmTJ+4cccsiuFStW\npLWW3BXCqL8HgZuAAlXdDfxjakXKPjZtojAR/7BMnz596xNPPHHQN77xjWFLliwpjRX2+OOPbwEY\nM2ZMc319fdH69euLvvzlLzcDVFdXN9fX1xfdcMMNG2fOnHnIBRdcMPSdd94pipf/p59+mj969Oid\nRUVFWllZufN///d/ewEcd9xxLQDFxcUaO4XsZ9PfN/k/uwD/WGzevDn/iiuuqHzssccaPL+RI0e2\n5OfnM2TIkF2ffPJJ/rp164rGjBnTDDB27Nhmv3TsWXaSTQHPLMg/gEmTJn26bt26dyoqKnY++eST\n/QCOPfbYZoCNGzf2OvTQQ3cWFBRw8skn+z6/Y445pgVg3LhxzevXry+qr68vOu6445oBTjrppOb1\n69cXXXzxxdvffvvtkvPOO2/YM8880zeeTH5lAOCoo45qAaioqNi1ffv2nJlnF0axqKq+nXJJsphB\ng9iZiH9Y+vfvv3v+/Pkb7r333o9uuOGGwYWFhRrUmfjWW2+VAKxZs6Z01KhRbcOHD2977bXXegOs\nXr26dNSoUW0jR45se/LJJxunTp269bbbbos7FLFfv37t9fX1hW1tbdLY2Fjk1ZK60wi5QQcM8n92\nAf5B7Nq1i29961vDbr/99r9FNotF3itVlREjRux8++23S8F5Ln5p2bPsJIMCnlmQvw8tLS17b0jf\nvn33lJaW7gHIy8tTgEMOOWT3Rx99VLh7926WL1/u+/z++te/lgC8/vrrpYcddljb6NGj2954443e\nAK+++mrpYYcd1lZYWKhz5sz56KmnnmqYFVWj6tWr157du3d3eDB+ZQD2L19hrzPThFEs20XkW0CR\niHwD2JximbKOmTP5uLiYDl/84mL2zJzJx2HTaGtrk/Hjx4967733Sk499dRRixYt6n3XXXcdfMIJ\nJ4w+99xzR0yePHnbkCFDdu3YsaPXOeecc9jmzZs7/J289957xSeddNKo1tbWvLPOOuuLadOmbXv6\n6acPqq6uHl1YWKhnnnnmF9dee23F2LFjR//kJz+p/Pa3v70jMn5FRcXOs88+e/ibb75Z7PnddNNN\nH1900UWHjR07dvSUKVO2RLY/dxdmnjrz4+JexR2fXa/iPTNPnRn62QE88sgjZW+//Xbvn/70p0PG\njRs3euHChb39wp1xxhlftLW1yUknnTRq3bp1xX5h7Fl2kpkzP6a447OkuHgPM8M/y9/97nf9xo4d\nO3rs2LGjt2zZ0uub3/zmZ5HnCwoKqKmpaTr++OMPf/rpp30H0hQUFOgpp5wy8sEHHxwwbdq0raed\ndlpzcXHxnurq6tFPPvlk2VVXXbX18ccfP7C6unp0dXX14RMnTmyKjH/OOed8evbZZ4949NFH944A\n9SsDYa8pGxHV2OVPRPoCPwOOBv4K3Kyqn6dBtrSxZs2ahmOPPXZbrDCzZ1M2axaDN22icNAgds6c\nycdhO+67ylVXXVVx6qmnfn7++ed3q/ueLma/Prts1suzBm/6+6bCQQcM2jnz1JkfJ9Jxn0zsWXaR\n2bPLmDVrMJs2FTJo0E5mzvw4bMd9MliwYEGfF154oc999933v+nKM4g1a9b0P/bYY4dmWg4/wowK\nu0BVf+YduLWXp1InUnYydSrb06VIjOQydezU7ZlSJEaSmTp1ezoVidE5wtRYFqnqGRHHv1PVC1Iu\nWRoJU2MxDMPIJrK5xhLYxyIiV4rIOuBEEVkrIutE5D3gs6A4EXH/IiJbReTnPuf+VURWiMgyEblf\nPNsEIotF5FV3e39XLqozxFOwhmEY2YL7vcraj1ZgU5iq3g/cLyLTVfXuBNO9FDgLONTn3H+q6q8A\nROQp4AzgJffcRFX9KMG8uoyIbGpubj7ArBsbhpELNDc3F4tI1g6kitvHoqp3i8jJwBBco2eq+ts4\ncT4KGuaoqusiDtsAb+imAk+IyE7gl6q6yC++iOynpc877zwGDRrkm19zczOlpTGniFBVVcUpp5yS\nlUMzVTUr5QpLd5W/vb19z4EHHphycyobN26sivYrKSkhP99/SkN3vd+5QrrkV1WWLl3Kr3/966TW\nWoK+l2vXrt1WV1d3cNh04ioWEXkQ6AuMA1biNJ/FVCxhEJHTgEOAl12viaq6TUSGAAtF5ISwo8/6\n9u3Lt7/9bd9zDQ0NDB06NG4aqbQR1BXCyp+tdFf5b7zxxjfr6upOSHX+Y8aM2e/DUVNTw6WXXuob\nvrve71whnfKPHz+e8ePHJzXNGOW9MaGEVDWmA5a528Xu9ql4cdxwlwA/Dzh3DPAq0D/g/FPAcWHy\nUVWmTJmiQdTV1QWeywVM/swSJP+ECRNWacjymWyX8+V9/nzVqipVEWc7f/7eUzkhfwy6q/yJlvcw\nEyQ9I3rtIjIQOCIhzRWFiIwAHgEuUtVtrp+482UQkT44c2YS05CGYWQ/tbUwZQo0NoKqs50yBX70\nIxg6lNPOOAOGDnXCJSOvoUMhLy95aRqhCDOP5SERKQHuAP6CoxRiIiIPAeNxZuufAPwC+Kqq3g7c\nAxwIzHXbIm8HXgDqRKQFKAB+oao2Vt0wuhszZkBzlAmu5maYPRtUnU5cT9kATJrUuXw8BebllYw0\njdDErLGISB5wtKq2qOrzqjpGQ4wQU9UfqOqXVHWEqp6vqm+5SgVV/brrP8F1f1bVXaparaonq+qX\nVfXJJF2fYRipoLO1gQ0b/P2jh/s3NztKqLP5BCmwGTPCxTe6REzFoqp7gOGugjEMwwhuzgrz0a+s\nDJ+Pl25n8glSYEH+1myWVMIojH7AmyLysIg86I4SM4zcxD4gwQTdm2j/adM6Xxu4+WaIHs4aNDw3\nPz9cPn5yBykwP/+uKErDlzB9LLemXArDSAfW7h5M0L1Ztgzmzu3oH0RQbcBLf8YMJ0xZGZSUwPbt\nzof+3HM75gGO8olWKn75BMl98cX+ad588/7pxWo26+nlopPErbGo6pJolw7BDCPphG13j/gDPvGi\ni3rGn2vQvXnwweAPfDRBtYToGkFTE7S0wLx50NAADzzg5FNVhYpAVdXe47j5BMn9H//hKK/ycsfP\nq/14/TaRJNpsZsTF+k6MzJHuZqkwH5Coj2Dx5s09o1kk6N6EnTjsVxvwnm9NTXyFPmkSNDSwZNEi\nR9lMmuTfbBadT6yPf1MTfPYZFBbuuw6/Zq5Ems2MUJhiMTJDOtu1vQ9ckKHRysrEPoK5SKQS79/f\ncWH6JIIoL3dqFJE1jMhmo8jnG0S8GsGkSftqLiJOniUlMHlyeLl37YKdUQtMNjc7z9lLI4wC6wo9\nsF8vlGIRkSNF5BQROVVETk21UEYPIJnDQf0+miLQq5eznTw5+ANXWuq08Xf1I5jN+DVFNTV1VOjn\nnrv/xzUeN98Me/Y42xkzOn44/Z5vNJEK3Y07YOHCjmHcmgzz5jnNZ8mQ2yOyjy1Sgfkpys7SUwcG\nxJuaj2MX7C/AQ657MJGp/elwOW/iIgbdVn4RVedV6+hEgk1+RPqXlzvOi+OXVjwXmUY8V1XVQfyc\nMulSVRXu+rz7G/b+lZaq/vCHzjbav5NxdxcVdTDxEvcaOiN3nGcby+RMTObP15aBAzvGiyV3MvJM\nMsky6RI/gGsjLJudKZbsJVD+oBeuvNz/Q+X3AeuqC5teael+L3q2KpZ3ZszY/wMVVvF25iOdn5/4\nfY+XT3l5+GsQ2Xfx8+fv/0wLClQLC2PLEy8Nn+ffIbx3HdEyxipfXckzhaRTsdyKM/s+Iy9RGGeK\nJUvx+4OLOOf7MgXVIDrzAUvCB7Fl4EDfFzwrFcv8+c4ff9h7GqREU6HE/T6WYRVevGsIqtV6/mGU\nZWdqGGHua6xylmieaSCdimUd8KG7XQesTSSDdDhTLCFId1U7zF9YUNNWql3Yppr583PLunGsj2ci\nzYWRHzyvvHRV4fuVuURqRn412VhlK2yZDFsu4tUwOlve4qWT5uaxtCmWXHCmWOLg9yJ4H5pkF9h4\nf4d+f2GJvqidcdHXG09G957klGKJpzy88/GUeORHNNYzClu78UsvIM09sdLoTNnyyzNWGkGKMjLt\nRPtz4t3voDyjn2eq3tkI0lljKcexbLwAuJOANVQy6UyxxCHeixCmDTlWR3pkk0NnPjJd6XiN9dH0\nXli/FzFku3ZOKZawnfTxwgd9oOOVhc588KPSbOvbN34aYfpb4hFLCccrF50ZLBLmvYh1HBQ+yUom\nnYplATAZGA58F3gukQzS4UyxxCHMixC2JhH0l9qZppZE5It2ngLpSjNBiObBnFIsfn0ssT6+ye40\nTkJ678yYET+NZPRJhBlllqgCJUaNC+I3G0Yqi0SVVpKUS8ZGhQFLEskgHc4USxzCFNREahLJ6EiP\nfIkSbb9PY6dmTikWjRgVFkupR5LsvrcupldXVxc/jWQoxM6M/opX6ygvd+5/IrWhoOeTqHJJ9J0I\nuMfpVCzPATXAYW7N5b9CxPkLsBWfpYmBYqAWWOpui13/ocAiYBlwfSIX0WMVS9iXOMxL4fen1lXl\nEdb5DQmNNUcijcMwc02x7JU3i4awJkLo9zUZCjGRNBLpk+vKnBvvBy/RfsdEmgFjlI10KpYBwN2u\ngrkbGBgizqEErHkPTAX+r7s/E5jq7j8BnOLuLwQOD3sRPVKxxGqmileN9xtv39XmrQAXt2nGc35z\nFyLlztDEsZxVLKoZv3edIeve17CKwGVvjSsZzXmx3tlY8eIRI++sHxUWQ7E8Dhzn7h8HPO7uvxcR\nZjpweUC6Gu16pGIJW6Pw6+RLZJhvV5RLVVX8ppnO/HGlkUwrFivvGSRsTT+CDjXGZDbnxfsxTOSn\nIcbgh2SV98D1WERknqpOFpF1boEGEEBVdVRQvBCUAzvc/U+AMnc/0m7ZJ8CgsAk2NzezePFi33MN\nDQ2B53KBIPlP27CBgOWROqLuo2tspP3SS6l/9122nHUWPPooAxYuZPQdd5Df1hYjutI2cCBFmzf7\n5hdZMDzai4qov+Yatpx1Fg0NDVBTEzef1gEDeC0Ln1M2lp+eWN4zwYlXX01xDHtn7UVF1NfUsCVC\n3r3yDx4Mjz7aMULkdQ0ezIDp0zlszhyKtmyhbcAAPrjsMrYMHsyAn/98f3/3nQUYsHChbzxC3rcT\nBwxwrHZHoaqccOGF/M/UqU5+XSERLZSII7jG8gQwxt0fg3+N5ScE1Fj8XI/8g+tsH0ii4/G98J2x\n7aU+f3DJ+ONKI5musfi5HlneM0Gs2nqCowhDk+q+sTAWA5JgwiiudWMReTLq+D+6pspYApzr7p/r\nHgOsEZHx7v7XgJe7mE/3xs/UdxgirfTGs9gbaTo81poVngXaPXv2raURjRdG1bFUmwpLsoaRTILK\nfFVVcDnvKola/U7EJL9nabmpKbYMSVgmIlCxiMihIvIVYISIjHfdacCR8RIVkYeAa4FLROT3IjJG\nRK51Tz8KHC0iS4Gj3WOAnwE3i8gyYKmqvtvpq+oJeGtV5OcnFi/yZYm1lkX0Bz+Za1aEUUSGkWlS\nvU6LH4msZpmoSf4wSxnEkyMksWos1cBlQAXwA9d9B2cWfkxU9Qeq+iVVHaGq56vqW6p6u3uuRVW/\nraqnuNtW1/8DVT1dVb+iqr/s0lV1ZyL/UGbMcApSdOEX6bj1iH4pgl6c+fP3X2MDUrdmhWFkI9EL\njaWjzCeymmWiS23HWm8orBwhCVQsqvoHVf0ecIaqfs91l6vqn7qUo9F5/P5Q5s6Fiy/uWPjnzQvX\n5BT04oD/nxBYTcPoWaS7dp1ILSnRpbaDiPcD2gkCR4VF8K8iopEeqvr9LuVqdI6gP5TnnnMKfTST\nJsV/EfzCDB0a/CdkysQwUof3fs2Y4SiIykrnI+/33lVW+isMVWcVVYjfn1Ja6vyYPvccbNhA64AB\nFN95Z5ff8zBLE88BHgYeAdYAf+9SjkYw8TriEml/7QrpyscwjP0JW0uKNYDHW346Fl4LxQMP7M3v\ntSeeSMrPY1zFoqrLXPeKqt4LDOtyrsb+BHTEdVgDPJH2166QrnwMw+g8kU3ZiZLKkW2EUCwi8p0I\n91Pg4JRI0tMJaOY6bM6cfcfpGqWSidEwhmEkjle7ie4niUUa3uUwTWEjXTcC+AI4L6US9VQCmpmK\ntmzZd5CuUSqZGA1jGEbnCduakKZ3OYxi2aCqN6rqLFX9d2BCSiXqqQQVDNWO/S3pGqVic00MI3eI\nN2Ham0aQpnc5jGKZHHX8z6kQpEfh10kfUDAEnP6WyZOd2kO82bWGYfQ8olsZyssdl6EWh1hGKK8E\nfgwMFpG1ON+4duDVNMnWPfE66b3+FE9pqDoFoaTEfzSH7jMmuXdOidUiDMPwCDO9IE3EmiB5v6qO\nBGao6ihVHamqh7uTJo1E8WopNTX7d9J7SqOpCVpa4qeVBFs+hmEYqSJWjeVLqvoOsCLCOCQAqro8\n5ZJ1J6JrKbFobnbsf7W3xw5nc0oMw8hSYs28/zrwDo6NsEgUMMWSCIkYfwNHqZSWxo5jc0oMw8hS\nAhWLqv6bu3u/qr7h+YvIUSmXqjtRW5uY8TdwOttcI5Da2IiI7GsuA5tTYhhGVhNmVFi0NeOZqRCk\nW+I1gcUiyACcO9x3SV2drV9iGEZOEauP5Z9wJkMeLiKuyVsKCLlksIhcAkzBaTq7MqrWcxX7JlpW\nAc+q6tUi8ihwLPApsFVVJyZ0NdlGrCaw0tJ9loTjGZzLotEehmEY8YjVx/I28BkwAPAmTuwE/jte\noiJyEM5Q5ROBwcA84GTvvKreBdzlhn0OeDoi+pWq+kr4S8hiYnWwR9Y6TGkYhtGNiDXcuFFVlwBX\nu9u3gOOB8hDpjsNZBXKnqn4I9BGRouhAIjIAGKaqr0V43yUiS0UkdydiekOLVf3PV1WZMjEMo9sS\npo/Fawb7BXAgMDdEnHJgR8TxJ0CZT7iLgKcijq9R1XHAN4DrROSw6AgiotEuhDzpI97COtbxbiRA\n1pd3w/AhzEJfXu9yf1WdLiJnhYizHUcJefRz/aKZBNR4B6q6zd1uF5EXcfpbPoiXWXNzM4sXL/Y9\n19DQEHguFZx49dUU+/SrKNA2cCAfXHYZWwYPhpAypVv+ZGPyJ59sKu/JxuTPLEmTX1VjOuA5YD4w\n3T1eGiLOQcAqnM7+SuAVnzCjgNei/A50t4XAMuBL8fJSVaZMmaJB1NXVBZ5LCSKqTiNYRyfSqeTS\nLn+S6a7yT5gwYZWGKJupcFlV3pOMyZ9ZklXew9RYLgAOV9U3RaQQp1M+nrLaISIPAEtwftanicgY\n4KuqersbrIZ9gwI8nhSRA1yFNF+dmf/ZS23tvhFdZW5LX1C/ik1oNAyjhxBGsfQFvisivwTWAr8K\nk7CqPoKznHEkb0Wc328+jKqeHSbtrCDaTEusZUCtX8UwjB5EmM77J4CVODWVlcDjKZUoVwhrpsUm\nNBqG0cMIU2NpVVVPmawXkYtTKVDOEMYIpIizsI5hGEYPItbMe8+i8SZ3rfuVwAlAQxrkyn4qK+Pb\nALN+FcMweiCxmsJ+wD7LxocD3wWOBPab6NgjCbMUqPWrGIbRA4ll3dgW9IqF12cSPSps+/Zgm1+G\nYRg9gLh9LCLy/3CGDO9FVb+fMolyCTMOaRiGsR9hOu/nuFsBqoHhqRPHMAzDyHXiKhZVXRZx+IqI\n/CmF8hiGYRg5TpimsO9EHA4BDk6dOIZhGEauE6YpbKS7VaCJfQt0GYZhGMZ+hJl5f6eq3gjcijOH\npS2lEmUj3voqeXnQv7/j8vIcv9poc2eGYRg9mzCK5Y/u9hfA+TgmXnoOkeurqDo2wZqanP3GRuec\nKRfDMIy9hFEs3lDjIar6A6A4hfJkH/FsgjU3O2EMwzAMIJxi6eVaNl7vHadQnuwjjE2wMGEMwzB6\nCGEUy7eA1cCtIlIC3JJakbKMMPa+zCaYYRjGXuIqFlXdrKr/qao7VbVFVZ8Pk7CIXCIiy0VkmYgc\n73PuQxFZ7LrBrv9QEVnkxrm+c5eUZMwmmGEYRkKEqbEkjIgchLN+ywSclSLv8wn2sKpOcN3Hrt9t\nwA2q+hXgDBE5PBXyhcIbCTZ5MpSUQHm5Ywa/vHzfvq21YhiGsR8pUSzAOGCpW8v5EOgjItFWkb8r\nIq+IyE0i4skxRlWXuvt/Bk6LTlhENNolXXq/kWAtLTBvHmzb5rg9e5y1VkypGCkkLeXdMJJMmJn3\n5cB1QDkwBfiWqv42TrRyYEfE8SdAGbDRPf4DMM/d/3/AJPc4LyrOoHjyATQ3N7N48WLfcw0NDYHn\ngjjx6qspjh4J1txM69VX89rgwQml1VU6I382YfInn2SX92zC5M8sSZNfVWM64E84s+1fdo8Xhohz\nDnBPxPFbQFFA2HOBe9399yL8fwJcHi8vVWXKlCkaRF1dXeC5/Zg/X7WqStWpp+zvRMKnlSQSkj8L\n6a7yT5gwYZWGKJupcEkr71mIyZ9ZklXewzSFlajqH4Hd7rGEiLMCOFlECkSkEvi7qu6dsS8iB0aE\nPQOod/fXRKxc+TXg5RB5JYfI5q8gbPSXYRhGXEKteS8i4wBE5Fjg7/EiqOoOEXkAWIIzwXKaiIwB\nvqqqtwPXishZOMqqHviZG/VnwMMiUgg8r6rvJnxFnaG2Fi6+GNrbg8PY6C/DMIxQhFEsU4HbgQE4\nH/4rwiSsqo8Aj0R5v+WemwHsN11dVT8ATg+TftLwaiqxlEpVla0IaRiGEZIwiiVfVb+dcknSTW2t\nY4olVtMXOEqloSEtIhmGYXQHwiiW34hIATAfeFpV4zaFZT1eLSWWDTCw5i/DMIxOEGbm/TnAZJwh\nxC+KyPyMXZMlAAARTUlEQVSUS5Vq4hmWBMjPt8mPhmEYnSDsBMlNwP8AjcCRqRMnTcQzGllaCnPn\nmlIxDMPoBHEVi4jchdPpfhbwb6p6fJwo2YtnpkVjTF42My2GYRhdIkwfyyLgWlWNMWwqB4jXr1Ja\nagrFMAwjCQQqFhE5U1VfAvoB/yyyb16kxjfpkn3E6lex4cSGYRhJI1aN5VB3OyLKPzeN4AX1q4jY\ncGLDMIwkEqhYVHWuu7tSI9ZgEZH0TmBMFpWV/nNWzEyLYRhGUgkzKuzaqOMfpUKQlOO3YJfNUzEM\nw0g6sfpYJgPfBcaIyAuudwHQkg7Bko7XfzJjhtMsVllp/SqGYRgpIFYfyx9xrAtfBdzl+u3EmdOS\nm0yaZIrEMAwjxQQ2hanqp6raqKrTgHagAhgGnJQu4QzDMIzcI8wKkjcA/4CjVBqAz4GzUyuWYRiG\nkauE6bw/V1W/AqwFvgJsS61IhmEYRi4TRrF41owFKARGh0lYRC4RkeUiskxEjo86968issI9d7+4\nsy9FZLGIvOpu70/kQgzDMIzsIIxJlwUiUgI8jLPa4+/iRRCRg4AfAycCg4F5wMkRQf5TVX/lhn0K\nZ3nil9xzE1X1o9BXYBiGYWQVYczm362qLar6mKoOVdWrQ6Q7DliqqjtV9UOgj4gURaS5LiJsG84S\nxeDM6n9CRBaJyBl+CYuIRrsQ8hhGTmLl3chFYs1jeYgA8y2qOiVOuuXAjojjT4AyYGNUHqcBh+AM\nawantrJNRIYAC0XkBFX9PE5eNDc3s3jxYt9zDQ0NgedyAZM/s2Sj/FbesxeT30VVfR1wWpALihMR\n9xzgnojjt4CiqDDHAK8C/QPSeAo4Ll5eqsqUKVM0iLq6usBzuYDJn1mC5J8wYcIqDVE2U+GsvGcv\n3VX+RMt7LFthS7x9EfkSUAU8DxwcQl+tAH7pLml8CPB3VW2LSG8E8Ahwgapuc/0E6KOqn4lIH+Bo\nnIXFDMMwjBwizEJfPwduAn7pej0WL46q7gAeAJYAjwM/EZExIuLZHbsHOBCY644A+0ecZrk6EXkF\nWAj8QlW3J3pBhmEYRny8dQ/z8pxtbW3y0g4zKuwfVPVUEalTVRWRwjAJq+ojOLWSSN5yz309IFp1\nmLQNwzCMxKmtdcwlNjY6K4Z4i+k2NjrrIE6fPoAJE7qeT6g170UkH1ARyQPyu56tYRiGkU68RXS9\n1UOiV2hvboY5cw5LSl5haiz3Aa/g9LHUAfcmJWfDMAwjbcRaRNdjy5ai2AFCElOxuDWULcDXcFaS\n/MD6PQzDMHKPoEV0IxkwoA0o7nJeMZvCVHUPcI2qfqKqq0ypGIZh5CbxFsstLYXLLvsgKXmF6WP5\nWER+KiKniMh4ERmflJwNwzCMlBE56qt/f9jmZz746Fr4yVC4IY+S64fC0b9NSt5h+liKgcNdB85s\n/OVJyd0wDMNIOl5Hvden0tS0f5jeJ9ay62tT2KlOoKbdjdyx9g6O+O8jmHR01xZEjKtYVPV7XcrB\nMAzDSAuRw4nj0XryDNq1Y29+2542Zrw0I/WKxTAMw8h+omsp8Wjv7d+bv+HTEL38cQg1j8UwDMPI\nTry+lJqa8EoFIP8L/978yn5xevlDYIrFMAwjR4me9BiKo2uR6UNpP6ARQTqcKsor4uYzb+6yXGHW\nvL8AuA7oh7OKpKrqqC7nbBiGYYTG6z/ZsAHKyhw/v055P8rL3fAVtcg3pqC9nKqNogiColT1q6Lm\nkJou969AuBrLTOA8VR2lqiNNqRiGYaQOP+OQkTUTVUehhFEqpaUwf74z1HjbNqj6/oy9SsXDUyoN\nP2ngrIFnJeUawiiW91R1Y/xguUH02O7+/R1jbL16OdtkW/k0DMMIIvp7dMABTl+Jp0AaG2Hy5MT7\nTwCqquDBB2GSWwGp/e9aGj/1bzNLRod9JGEUS7mIvC4iD4nIgyLyYFIlSAPewxNxHpKf1m9vd7be\ngzQlYxhGNEG1Cc/vootO3PvN8PuJjf6hjf4effHF/nlGG4uMh1dLaWjoqFSm/Cl44d9kdNhHEkax\n3ARcA8wHal2XMyxcOCCmRU8/ok1JZ6tySeV6CobhkYlyFu+jHL0fKVfkj6TXEhEvbJh8ohVBY6NT\nk4isYWzeXLz3x9TvJzb6hzZRpRGPqiq4+I5aZmwdSt6NefT/VX/6/6o/Nc/W0LzLv8pTWlCalA77\nSMJMkFzims0fAFFDCGIgIpcAU3Bm6l+pqm9EnCsGHgYqgQ3AparaKiJDcdZwKQL+rKq3hL6SKGpr\nYdqcWv6h+Me8U7DdyahvHteftYfHh7s9WaXbodntBSttgj35kNcOzfvONzeXUbMGatZth5Yy52pK\nE92PSLulPME0DoLnJDjMJU7ajXnt1Kwpp2ZNZ+RLhdwR+8+nMO0Uyl1eUMnUETVMYEJni2Ha8Mp7\n05jrYfHfknQfM1HOyvaV928572ZTxHsatN9Yup2aNWX75HLjtrtpN0W810Fhw+SjIb4Z6hs20X3/\ntAPjtZRxwAHQuKeJ2VuczniAppb4HTEP/tODSemwj0Q0jsoUkcuAa3GWJP4M2KaqJ8SJcxDwEnAi\nMBiYp6onR5yfChysqjeJyExgi6rOFpEngF+r6lIRWQj8i6q+F+8iLr/8cv3Nb36z97i2Fr53dy0X\nDv4eDz2/i9679oX9ogB+8E/w+DHxUjUM6KXFPHrBnP1evNNPP311XV1dzPcgVQSV911nT4HCBBvi\njR6N12nvsXjxYib4rPSVaHkP0xQ2FWf9+beB0cCaEHHGAUtVdaeqfgj0EZFIQ/+nAQvc/T+5xwBj\nVHWpu//nCP+9iIhGu+gwM2bArlNmcMuSjkoFoPcuuOWlEFdgGMBuaWXGSzMyln8i5d2UipEIqWgC\n8whj0uVTVd3prs2yk33GKGNRDuyIOP4EKAM2+pz3zkFHRfcJMChEXjQ3N7N48eK9xxs2nAb9NlD5\nqX/4IH/D8GPDpxs6lK9ME1TeDSMsA4sGctmwyxjcNLhDWWpoaEhKWQ+jWF4XkRLgj8B/A++EiLMd\nODDiuJ/r53c+8tyeGHEAUNX9+nkuv/xyjay+VVZC46eVbOjXyFAfJbKhX4grMAyXyn6V+zUP3Hjj\njWnJO5HyzoGJTL82eiKlBaUx+1SCmsISLe9xm8JU9TpVbVHVO4CTVfWfQ6S7AjhZRApEpBL4u6q2\nRZxfApzr7p/rHgOsiVjv5WvAy6GuIoqbb4aCpTdz/WkFfFHQ8dwXBXD9mZ1J1eiJ9NLilDUXJAuv\nvLOzNNOiGFmCZ6qlvKSc8pJyBKGqX1VKOur9CGPS5TDg34C+wD+KyFWqelesOKq6Q0QewFEYCkwT\nkTHAV1X1duBR4BERWQp8BHim+X8GPCwihcDzqvpuZy7KGbs9iWlz4Adn/phbXt1O5afw8YF5XHfG\nHl74cjnlwPaW7ZSVuCM/WprIl3zatZ3yknLXbzvSWoa2k8FRSgeBxhgVVtpEnuSzh3aktTwJsubW\nyK10jApLx4vYFSLLe9OY66FfCkaFpa2cle0t773zyigu7vieBu1Hv6te3DDvdSL5xPtmJJpGKtKu\n7FfJzWfenNlyq6oxHbAQOAZY5B6/FC9Out2UKVM0iLq6usBzuYDJn1mC5J8wYcIqtfKedEz+zJKs\n8h6mj0VV9W2R0FNY0s7atWu3nX766fs1MH/++efVAH369Fmdfqm6jsmfWeLIX5VmcfZi5T076eby\nJ1Tew8xjeRL4HTAN+BXwz6r6nUQyyRTe0Ez16QDNBUz+zJJr8ueavNGY/JklmfKHmcfyA+A4nOHB\nJ+HMazEMwzAMX+LWWHIZ+4PILCZ/esk1eaMx+TNLMuUP7GMRkQ+A3dHeTr62JothGIbhT6zO+zrg\nUOA54AlV3ZwekZJHrv45eJj8mSXX5M81eaMx+TNLMuWP2RTmzif5OvAtoBR4QFX/K1mZG4ZhGN2P\nmJ33qroTZ/b7Uhx7XkemQyjDMAwjd4nVxzIJuBDYBTwNnKkdzbIYhmEYxn7EqrHMA4YAB+Es2PUn\nEXlBRF5Ii2RJQEQuEZHlIrJMRI7PtDxhEJEWEVnsukvF4X4RWSoiC0SkLH4q6UNE/iIiW0Xk5+6x\nr7wiUuYeL3XPZ0V7tI/8E0RkY8QzqHb9s1X+b4rIuyLSGuW/WERedbf3R/hn1TsRQ/7jXRmXu4sG\nev5ZJX8kIvIL91q8spPv+vteS7aR1HsbNCUfZ6alr0tkan+mHI5CfAMoBIYBr2RappByvx91fA7w\nsLv/XeC2TMsYJd+hwCXAz2PJC9wGTHb3HwHOybTsAfJPAOb4hMtW+cuBYp9ysxg4NMov696JGPIv\nc2UsdGU+KBvlj5L5F0CNj/9+15JpWX1kTOq9DayxqGpjkAuKk2XEW2wsWxkkIktE5FlxlmoOWhQt\nK1DVj6K8guTNyuvwkR/g7IiaSYnrl63yN6lqq98p4AkRWSQiZ7h+WfdO+MnvytRbVT9Up593KY7s\nWSe/D/8qIq+IyI8h5rVkG0m9t2Fm3ucqQYuNZTtDVfU04DfAw+y/KNpBmRIsJEHylrnHnn+2PovV\nwEhVPQVnKe5rXP9ckd9jojrLgV8M/IeI9CF33oly9t1r2Cdntst/P3As8FXgPBE5heBryTaSem/D\nGKHMVeItNpaVqOo2d/sXEfk18DodF0XbERQ3S4hexM2Td4d7/AlZ/CxU9fOIw1rgVnc/Y/K7bfXL\nfE79WVVv8osTUY7+JiJrgBFk6J3ohPxBckqAf9oIeS0tIvIscAId31/I3rKf1LLRnRXLCuCXIlIA\nHML+i41lHSJyANCiqu0icgywDWdNm28Cv6fjomjZSpC83uJuv3W3z2ZEujiISD9V9dYdPQOod/cz\nJr+qtgMnhg3vDizoo6qfuTWVo4FGoIEMvBOJyq+qrSLyhTiLBG4ETga8JQwz+k7HuhYROVBVP3Hv\n/wTg0TjXkk0k9XvZbRWL+iw2lmGRwnAk8BsR+RxH5stxloP+ujiLon2G0yGeNYjIQ8B4oEhETgD+\nD/7y/gp4TER+CLwNZMXoQh/5XxCR7wPNOIr9+27QbJX/FOAGoEJEFgIP4PQB1YlIC1AA/EJVt7vh\ns+qd8JNfVZ/Fke1xnFrKA6q6ww2fVfJHcY+IjMaRebGqPuf6+15LNpHs72W3NkJpGIZhpJ/u3Hlv\nGIZhZABTLIZhGEZSMcViGIZhJBVTLIZhGEZSMcViGIZhJBVTLFmKiAwVERWR8yP83heR0yKM3EUa\nrByRxLzHiMi17v757hh871xtsvIxDA8r790LG26cpbh2wp4HPgVOUlUVkfdVdUREmA7HKZLjURyj\njK+kMh+jZ2PlvXthNZbs5mMci6PfSCSS+/f3hog8KSKrRGSa6z9QRJ53jVw+JyIHi0hphN9iERkl\njun4OSJyJI614vtF5Gk3jfeD0vLOi8i/uf5PJPNmGN0eK+/dBFMs2c8twHUiCa//MQS4DDgJ+J6I\nDAB+BjzuGrl8wj0+HNihqqep6gTgfS8BVf0f4L+AK1V1YlT6fmmBY83B8y8TkaMSlNvo2Vh57waY\nYslyXLPuq4Hz44WN4j1V/VxVdwF/xVljYTSw3D2/HOclexNYLSLzReReoG/I9P3SAtitqm+5+xtw\nrKYaRiisvHcPTLHkBrcC1yUY53AROUBEegFHAR/iGFQc754f7x4XAXepag2wFZgclc5O/G3K+aXl\nR1astGjkFFbec5xua4SyO6GqH4nI6zjtv2FpAB4CRgJzVXWLiNwGzBWRy3CMLH4Xx/DlfSKyG+dH\n42KclUI9FgCzRORdVb08wt8vLcPoMlbecx8bFdYNcUfYzFHVszIsimGkHCvv2Yc1hRmGYRhJxWos\nhmEYRlKxGothGIaRVEyxGIZhGEnFFIthGIaRVEyxGIZhGEnFFIthGIaRVEyxGIZhGEnl/wNcGa5d\nlnVvlAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x109d0f9b0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(6,2.5))\n",
    "ax1 = fig.add_subplot(121)\n",
    "l1, = ax1.plot(ivals[0::3], plot_vals_hyphy[0::3], 'bo', label='1st codon position')\n",
    "l2, = ax1.plot(ivals[1::3], plot_vals_hyphy[1::3], 'go', label='2nd codon position')\n",
    "l3, = ax1.plot(ivals[2::3], plot_vals_hyphy[2::3], 'ro', label='3rd codon position')\n",
    "ax1.set_ylabel('Mean relative substitution rate')\n",
    "ax1.set_xlabel('NT position')\n",
    "ax2 = fig.add_subplot(122, sharey=ax1)\n",
    "ax2.plot(ivals_r[0::3], plot_vals_hyphy_r[0::3], 'bo')\n",
    "ax2.plot(ivals_r[1::3], plot_vals_hyphy_r[1::3], 'go')\n",
    "ax2.plot(ivals_r[2::3], plot_vals_hyphy_r[2::3], 'ro')\n",
    "ax2.set_xlabel('NT position')\n",
    "plt.setp(ax2.get_yticklabels(), visible=False);\n",
    "fig.subplots_adjust(top=0.85, left=0.1, right=0.9, bottom=0.18)  # create some space below the plots by increasing the bottom-value\n",
    "plt.figlegend((l1, l2, l3), ('1st position', '2nd position', '3rd position'), bbox_to_anchor=(0.15, 0.88, 0.7, 0.85), loc=3,\n",
    "           ncol=3, mode=\"expand\", borderaxespad=0, fontsize=8)\n",
    "# plt.tight_layout()\n",
    "plt.savefig('../Results/Figures/position_dependent_rates.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Defining a SD threshold strength"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with open('../Data/json-energyRef-CCUCCU.txt') as infile:\n",
    "    fragment_energy_dict = json.load(infile)\n",
    "asd_seq = 'CCTCCT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "real_utrs = []\n",
    "include = 20\n",
    "genome = list(SeqIO.parse('/Users/adamhockenberry/Projects/2017/Genome_database/Data/GBFs/511145.12.PATRIC.gbf', 'genbank'))\n",
    "for record in genome:\n",
    "    for feature in record.features[:]:\n",
    "        if feature.type == 'CDS':\n",
    "            if feature.location.strand == 1:\n",
    "                cds = str(record.seq[feature.location.start:feature.location.end])\n",
    "            elif feature.location.strand == -1:\n",
    "                cds = str(record.seq[feature.location.start:feature.location.end].reverse_complement())\n",
    "            if len(cds)%3 != 0 or len(cds) < 100 or cds[-3:] not in ['TAA', 'TGA', 'TAG']:\n",
    "                continue\n",
    "            if feature.location.strand == 1:\n",
    "                real_utrs.append(record.seq[feature.location.start-include:feature.location.start])\n",
    "            elif feature.location.strand == -1:\n",
    "                real_utrs.append(record.seq[feature.location.end:feature.location.end+include].reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4525\n",
      "90500\n"
     ]
    }
   ],
   "source": [
    "randomizations = 20\n",
    "real_utrs = [str(i) for i in real_utrs]\n",
    "strongest_real = []\n",
    "strongest_fake = []\n",
    "strongest_fake_dict = {}\n",
    "for randomization in range(randomizations):\n",
    "    strongest_fake_dict[randomization] = []\n",
    "\n",
    "\n",
    "for utr in real_utrs[:]:\n",
    "    energy_list = []\n",
    "    for i in range(0, len(utr)-len(asd_seq)-4):\n",
    "        fragment = utr[i:i+len(asd_seq)]\n",
    "        energy = fragment_energy_dict[fragment.replace('T', 'U')]\n",
    "        energy_list.append(energy)\n",
    "    strongest_real.append(min(energy_list))\n",
    "    \n",
    "    for randomization in range(randomizations):\n",
    "        energy_list = []\n",
    "        utr_rand = np.array(list(utr))\n",
    "        np.random.shuffle(utr_rand)\n",
    "        utr_rand = ''.join(list(utr_rand))\n",
    "        for i in range(0, len(utr_rand)-len(asd_seq)-4):\n",
    "            fragment = utr_rand[i:i+len(asd_seq)]\n",
    "            energy = fragment_energy_dict[fragment.replace('T', 'U')]\n",
    "            energy_list.append(energy)\n",
    "        strongest_fake.append(min(energy_list))\n",
    "        strongest_fake_dict[randomization].append(min(energy_list))\n",
    "print(len(strongest_real))\n",
    "print(len(strongest_fake))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAANUAAACPCAYAAABtVZ4SAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAHXNJREFUeJztnXl8VdW1+L8rIWFMDQIKQRRkCKJAEHLBOACtA1UsfRRF\nHKpFeeJ7LQhFH9JKpfpRf+JEqtY6PSkoaKlPNFpUZJQgoaIIVgJRUosESaAomEAIWb8/9rlwubnD\nSXJHsr+fz/mcc/bZw7rnnnX2cNbeS1QVi8USOVLiLYDFcqJhlcpiiTBWqSyWCGOVymKJMFapLJYI\n0yzeAgRiw4YNP01NTX0ESI23LJYmTS2wuaam5paBAwfudpsoIZWqWbNm07t3717VunXrg/GWxdJ0\nqa2tlfLy8j67du16DviJ23QJ2fxT1Y6tWrWyCmWJKykpKdqhQ4dvgXPqlS5K8jQaEYm3CBYLKSkp\nSj31JGGVKtrs3bs3ZdiwYT08Hk92//79e69atapVcXFx+qhRo7q5ST927NgzBg4cmL19+/a0YcOG\n9Rg0aFD2Y4891v7RRx9tHy5tQUFBxtSpU7N8w6ZOnZr1+uuvZ3jPPR5P9ltvvdXG4/Fk9+7du09W\nVlZfj8eTfe+9957i8Xiyc3Nzs3Nycno//PDDYctLNAoKCjKysrL6Dh48uFdeXl7PXbt21avv7PF4\nsg8fPnxc2MCBA7N98586dWrWmDFjuno8nuyMjIyc3NzcbI/Hkz1//vzMxpTthoTsU/kyd+7cHo1J\nf+ONN5YECn/66afbjRo1at+UKVMqDh8+zIEDB1J2797t+n6UlJS0+Oijj4pLS0vT2rRpU7tixYqS\n/Pz8djU1NRGrYq+44ooDV1xxRXFBQUHGu+++m5Gfn78T4M0332xbWFhYnJKSwqBBg3pPmzatIlJl\nehn3zIeNuu8L/nNIwPvuZcyYMXvy8/N3PvXUUye/8MILJ8+YMaO8MeUFYtGiRaVglHDNmjXFaWlp\nFBQUZES77CZbU7Vq1ar2ww8/bF1WVtYsLS2Ntm3b1gLs2LEj/bLLLut+9tlnn/XFF1+kFRQUZEya\nNCkLID8/v11+fn67SZMmZW3durXl8OHDe0yePPm0tWvXZowdO/YM3/ynTZvWyePxZA8ZMqRXcXFx\nOsBVV13VNS8vr9ezzz4bkdqlqqpKDh06JADvvfde6379+vUePHhwr8ceeyxpaq99+/YdrSmWLVvW\n2uPxZJ977rm958yZ0w7grrvu6pibm5vdr1+/3mvWrGkZrbJHjx7dNTc3N3vw4MG9jhw50qh8E76m\niha33Xbb3h07dqQPHTq0V/v27WsWLlz4JUBlZWXq22+/XfzMM8+c/PLLL7cdMGBAlX/a/Pz8nWvW\nrMlYvnx5SXFxcfqdd97JK6+88s/8/Px2AOvWrWu5c+fO9KKiouINGza0mDVrVqcJEyaUp6amamFh\n4dbp06d3rK6ubtQLLS8vL7ukpKTFjBkzvgYoKCg46f777/965MiR+2traxuTdUxYtGhRu2XLlp1U\nVVWVUlRUtAVg5syZWUuWLNmWmZlZe8EFF/S69dZb986YMWP3Aw88sGvz5s3NZ8yYkfXGG29sj3TZ\nhw4dkrKysvT169cX19bWkpLSuLqmydZUzZs314cffrhs69at/7jxxhsrHnzwwVMBevbsWZWamkqX\nLl0O79u3L1VEjprxu7Xo37RpU4vCwsIMj8eTPXHixDP279+fsm3btuY5OTmVALm5uZX+aVq0aFF7\n8OBB1/9HYWFh8fvvv1+8evXqDIApU6aUL1y4sO2oUaO6rVy5spXbfOLFmDFj9mzevPnznJyc77/4\n4ot0gC1btrQaMWJEzyFDhmSXl5enlZWVNXv66adPHjRoUPb48eO7fvPNN2lu8j548KC0bNky6JvF\nv+zmzZvrddddt2fUqFHdJk+e3LmxNVWTVaqtW7eme5tOHTt2PFxbWytw/Kijqkrbtm2P7Nq1Kw1g\n06ZNrpofffr0OTh06NBvi4qKiouKiooXLVpU2qNHj+pPP/20FcBHH31U56E/55xzqtasWdMGoLy8\nPDU1NTWsBp977rkHVZUNGza0aN++fc38+fO/mjNnzo7f/e53nd3ImQjMnDmz7N577+0I0KdPn8ql\nS5duKyoqKt68efM/unXrdvi55547Zd26dcXPP/98qaoG7a+mpqZSXl6eCrBq1ao2ffv2rdPCCFZ2\nTU0NEyZM2Lt48eLtFRUVzVauXNm6Mb+pyTb/ioqKWo0ZM6ZjixYttFmzZrXz5s0rDTTIMHjw4Kpd\nu3alDx06tEfbtm2PAGH/rLy8vKrFixcf9ng82SKiV1111d5p06ZVPPnkkx3OO++8Xqeddlp1ly5d\nqn3TjB079tvFixdnejye7CNHjvDQQw/tcPM7JkyYUJGfn9/hzDPPPPTGG2+0raysTJkyZcou1zci\nzvTv3//Qnj170r766qtms2bN+vrSSy/toapkZmYeeeedd74YMGDA9x6PJ/u88847ECqf++67b8eI\nESN6pKam0r1794NXX331t27L3rlzZ7PRo0d3r62tlTZt2hzJzc0N+x+HQhJxkuLGjRtL+/fvH/ER\nLYulIWzcuLF9//79u7qN32SbfxZLtLBKZbFEmIRVqkRsllqaHs4AVr2+USSkUonIrsrKyhbxlsPS\ntHGs1E8CNtcnXUKO/tXU1DxYUlLyCNAm3rJYmjRH51PVJ1FCjv5ZLMlMQjb/LJZkJqxSicj5sRDE\nYjlRCNv8E5HngL7AK8BcVd0TC8ECMXLkSO3cOXEscCorK2nVKnHM7ILKU1xs9tnZda/FQ544EUqe\nrVu3VixfvrxDRApS1bAb0Br4BfA+Rrl+5CZdpLdx48ZpIrF8+fLYFlhdbbYgBJVn6FCzNYTycrM1\ngJjfnzCEkmfYsGF/1wg9p277VK2BU5x9GTBSRBZFRKst7rnkErPFkjFjzGZxTdghdRF5HTO0/b/A\nUFU95IS/GF3RLHW4pV4ju5Y44eY71UOqWug9EZFTVHW3qt4UPbEsAbn++nhLkLAsWLDg6PG4cePi\nKIm7IfX7/M6fiIYgFhdUVprNktAEralE5ELgIuAMEZnhBKcBiTP81tS4/HKzX7EirmJYQhOq+fcd\nsAM4AHzthFVj+laWeHDbbU2jzCQnqFKp6kZgIzA3duJYQjJ2bNMoM8kJ1fybp6o3iMg2wPuFWABV\n1V4xkc5yPN86M8RPOil2Zf7rX2bfpUvsykxyQtVUNzj7nrETxxKSUaPMPpZ9qhtuiH2ZSY4b27+r\nnP1ZIvIXEbks+mJZAjJpktksCY2b71QTgb8AdwCPA7OBd6IplCUIo0fHW4KkwPvNKl7fq9x8p2oh\nIilAqqquAQ5FWSZLMCoqzGZJaNzUVGsxo4DTRcROcY8nXhs8279JaMIqlapOA6Z5z0Xk4qhKZAnO\nr38dsaxufnE9AM/flBuzMpsKbgxqzwcmA+0wzUUFfhhluSyBuPLKplFmhIhX38pN8++PwK3AV/XJ\nWERuAv4To4S/UtUNAeLMAq5T1Ub5Qmoy7HJWc+7YMXZlxmmCYzLjZqBim6quVdWvvVu4BCLSFpgE\nDAOuB/IDxDkVCPoRWUTUf3Mh64nNNdeYLZbceqvZLK5xU1OliMhrwEc4lhWqen+YNB5gtapWA9tF\nJENEmnvnYjncDTwAvOZW2MOHD7MigTrppaWlMZXnZMegdm+QMoPJk7NvHwCf+Fyr2GP8lIeTP1Ba\nt8Ty/lSEGBX1yhAredwo1esNyLcd8G+f833AyZhZw4hIT6CNqn4azGG2BnCbcu211+qwYcMaIE50\nWLFiBTGVJ0xZQeXJzHSSH7s2r3S9ExZmoCJAWrfE8v6UlZUFveaVIZQ8s2bNipgsbkb/5jrNuS6O\nErhpMu4FMn3OT3LCvNwDzKyPoBasHV6S4MZMaTxQAMwTkWbAGy7yXQdcICJpInI6cMCv6Xcm8KSI\nLAE6iUidPpclADfccMwWz5KwuGn+3QxcCLyvqjUiEnbNKVX9t4g8BazE9MMmi0gOcImqzlbV87xx\nRaREVa1Bmxt++9umUWaS40apjqhqrc/oW2rI2A6q+gLwgl/wJwHi2eF0t1wch+/u8SgzyXGjVK+K\nSAHQ1VmWbGGUZbIE48svzf7MMyOWZVjLik+c92BOTsTKjCS+C74kCm4GKp4QkaXAOcDnqvpZ9MWy\nBGT8eLOP5WeF22+PfZlJTqiZvxf5Be0G2onIRaq6KrpiWQISwWHfpoS3NuvUqVNMygtVU3mHmc4E\nOgIbgBzMYjBWqeLB0KHxlsDiglDT6ScAiMhfgcuckb9mmLXULfHA2uElBW4GKs4E0oEaZx+5XvIJ\niLfjDy6mVdQXrw2e7d8kNG6U6n7gYxHZBZyKtYSIH/eHM7k8QcpMctyM/v3FaQK2BypUtV6eui0R\nJC+vaZSZ5LhypO0o0u4oy2IJx2bHSfo558SuzELHN4VVLtckpHf6ExXXU9iD8ctfmn0s+1QzZsS+\nzCTHzXT6lqpaFQthLGGYPTveElhc4Kam+kBE1gHPBZoSb4khuREeTbREhbBTP1R1IObb1GQRWS0i\nE0XkB9EXzVKHTz45ZotnSVjcDlSsFJGDwFTMirXXicibqvpQVKVrQrjqb1k7vKTATZ/qdmAssBl4\nRFWLnPClgFWqWPL4402jzCTHTU1VBVyqqvv9wu0U1FgTj+kXCTrlI5Fxs97ED30VSkT+CKCqwVfa\nsESH9evNFkuWLjVbA9i0aVNCzneKNqGmfpwGnAH0EBHvl780oE8sBLME4I47zD6Wfar7HD/qdgaw\na0I1/wYCPwWygAlOWDXwcLSFsgThiSfiLUG9+L89p5K+P534OLSJH6GmfiwGFovIWar6eQxlsgQj\nluZJlgYTtE8lIl53D1NF5BnfLUayWfwpLDxmixdDvtm9u0n2jRpKqOaft3c6PxaCWFxg7fCSglDN\nv43OfmVDMg7l9UNE7gR+hpn4uAGYpKrWAUE4/vSno4eNNs6tR5nrCwqiW8YJRqjm3zYR2eq3bROR\nreEydeH14/9UdbCqno+Z+FjH35X1+hGA7OzYT6XPzmZ/VlajsliwYEGTaj6Gqql6NiLfkF4/VHWb\nT9xDmBorLMng9cPrTQPqetQI5Wkj0LXHPjJhUwYar7AnOXZ/3+bkBIzfEK8fwWT10q6wkDalpZT2\n7Rvy3m/atAmAvn37Hg07cqSW6urqox45ovHfhfL24c+hQ4fi6/VDRH6kqu+LyLX+11T15TD5hvT6\n4VPGUKATAVZnSlavH15vGlDXo0YoTxuBrtUJu+ces7/99oDxG+L1I5isYGqYs+bP5xTgwPDhx6X3\n91Lo9brhGyd1xaukp6fTvn37OteC5VNfQnn78KdTp05x9/pxmrNvSI0VzusHItIPeBC40vanXPKC\n/yralkQkVPNvrrOfBSAiHVS13GW+64D7RCQNUxMd5/VDRHpg1ln/maq6r7+bOhFc7tkNc7e35OyD\nKUePm9pH3Ibixkr9Z5gVlb5xXIr+RlUXhUoTzusH8DimJpvrOH2brapvNe6nJA++y5jVC68NnjUZ\nSmjcWKn/BhikqvtFJAOjKCGVCkJ7/VDVkfUV1EJE7fC+3mlcN3fO6tyofJJhVG/u9pYATI/Nqs+u\nlKoUOOAcf++cW6JEyFps3ryg8aP1veqhMbcfPU4GBUoEQo3+PYtpumUCG0XkY2AAENY7faISsw+m\n0SIKbkm9NRYEvicVme0jXma08dZMN3aLz3pFoWoqa56UIBx9GXTcYwJGjIhZ2Rd9+gEAq/pdELMy\nk51Qo39HzZNEJBU4BQjsSt4SGx580OxjqFQji5YAkVeqSDUl410rBcLN6N8twB1AB+A7oAIYFGW5\nLIFYeOI7sfRVtoZ+EI43bgYqJgJ9gXeBy4CnoiqRJTgdO8ZbgqB4a4yGqkE0BkG8MsUaN2tUfOvY\n8KVgZv72jq5IlqC8+abZLAmNm5pqvYi0BN4ANgHW52+8eOQRs7/yyrgU7/vmT6Q+TKLhxpXOdOfw\nYRF5TlX3RVkmSzAWhf3mHpAtu8xiWA1pYtw37s4GlenLUWVsIt+53AxU9AYewBjWbhORu1R1S9Ql\ns9SlfeO+GTXEPOq71naF7/ripvn3v8CvMUayHmAuMDiaQlkM/qZET/7SLAi8YdDwOnFvfnE9FXsO\nEumZMZdsWAbAe+ceP480XoMAyYAbpdqjqt7VRtaKyJ5oCtRYoupzN878aKnxYR5IqaJFMKUKxvGj\neKmNKjvUXKtj5RjlTiQlD2Wm5J2c+L2zKm0R5vuU2+kflgZwzGyoLk9Miv6SiyPuf71R6Y9/uKsb\nJ0ySEqqm8k5O/IezPx3jotS6KY0TVa3a1AnzVcLq6sAPcXX1oYDh8cS/ZjmRRhNDmSkdN79YRNrb\nCYXxJXfdewCsH3xJg9KHqgUThVDD9sliJe9m9O/HmEmF34jIKcBUVX076pIlKcc/uLkBrzV0DtOw\n5X8FjFIlg4LUh0TqEzUWNwMVMwGPqn4rIpnAEsAqVRyYMyW8r6hID9Tc/fO7G51HU8ONUqmqfusc\n7DvR1t+L9xwrNzVOpGbphsJ/NM3LofTmUSszHI21J4wXbpRqiYi8C3wIDAHeia5I0SPchLzYy1E/\nst9+CYBlOcNcxfe+MCY3qDTDyA//BkDBkB83IpemRUilErMqy8vAYiAbeE1VrSfnODHi72bhF7dK\nFQku2rwGsEpVH0IqlaqqiPxBVX8MbIyRTA3C+1YONVDgH9dtnonyEfmuX9wTbxEaRUMHI5Jl1M+L\nm+bfZyIyFigEjgCo6s6oStUAojEalijNRS9HUsP/Xb73IZp9sHiQLFbybpRqoLNNdM6VAA4F/Anj\n9aMF8Dzmg/JXwM2qejBQPtEm0RQnFPU1GarPiyaRh7QDyZbI8obrU6UAV9djZVpvOq/XjyFAZ2Ae\n4LvIwU3AFlW9TkRmOudP++VRZ5QxUtOrQz1soZqGXhMebw1Qsefg0TXJY9FErK9SWeKDBFvGXETG\nYVamLcOsTzFBVVe4ylTkMuByVZ3snG/EfOs65JwvAB5S1Y9FZABwp6qO88ujjmB5eXmkp6d/5PbH\nRZP9+/cPBMjIyLDyBCAJ5Tlj+fLlHSJRVqiaagrQz1mZtjNmyscKl/mG8/rhe9177Tj8vX6IiBYW\nFqKqCbHojFfprTyBacryhFqj4oCq7ncE+Zr6LU8WzuuH7/U6HkEslmQmVE2V43z0BaNQA7znqnpp\nmHxDev3ArMd+OWZt9cudc4vlhCBUn+qMYIlU9Z9hMxYZD9yC4/UD4y3xElWd7Swk8wLGB9YO4Bfh\nRv98qu+EWNDTyhOapixPUKWyWCwNw826fxaLpR5YpbJYIoxVKoslwlilslgiTEIqlYj8h4h8LiIH\n/cLPFZE1IlLo2BYGSjvDibNMRLpGQbaHRGSFs5WJyK8CxNnuE+c3kZbBr6x7nHvlLa/OumAiMkJE\n1jrbZVGW53ER+dDZpgeJU+Uj781RkuMm5zlZIyLn+l1rISIvichqZ98iooWrasJtGIuLFkCJX/ga\noBuQDmwA2vpd7w0sdY4vAhZGWc5NQFaA8JJolutX1j3A9SGupwKfYj62Z2Km8KRGUZ6ezj4FWAt0\nj/X9Ado6z0e687x84Hd9InC3czwTmBjJ8hOyplLVPer33UpEmgOtVXW7Gi8kqzEr5voyFHjLyWMV\n0D9aMjpvv2808DQYEZHlIrJERHKiJYMPd4rIByIyKcC1HsB2Vd2nZh38UicsKqjqNmdfi/k2eSRA\ntI4islJEXotGawLzXKxW1WpV3Q5kOM+Pl6FAgXP8pnMeMRJSqYLQDmMn6CWQzaC/zWHjlkgNzfXA\nS0GuDVbV4cD/EH03r3/AvDwuAX4iIhf6XQ9mhxlVROQ64EtVLQ1wuauqDgX+hJkCFGnC/eawtqeN\nwc18qqjgtP3XBLj0lqreGyA8nD1hoDiB3pKNls25PgrT9KqDOusjqupGEakUkbaq+u9AcSMhj3Nc\nJSKvYVYRXu0Tx819i6g8InIx8AsgoM8fn/vzjog82RhZghBf29Notm0j0DYO1Kc6HUgD/k7dPtVZ\nwDvOcR5R6lMBlwKvBLnWHGjhHHcGtuJYrkRJlkxnL8CrmCk3vtdTMf2oHzhbtPtUgzF9qcwg19t4\nywf6AR9GQYa2zvOR5jwv/n2q24AZzvEMItynisqNjcBNuRBYClQ6+9FO+CBHsQqB8T7xX/I5vtuJ\nsxw4M0ry/Rm40i/sccy8sy6YTvJqR87hUb5XLzoP8YfAg/7yOMeXO3HW+itdFOTZ7GwrnG2g73+E\n6e98DKzCGFL3j5Ic4537v8Z5bnKAO5xrLYEFzn+0wPsSjNRmbf8slgiTTAMVFktSYJXKYokwVqks\nlghjlcpiiTBWqSyWCBNXpRKRYSLynM/5aY6R5WRn/4ljtLpCRAqc+N7zImflXP88u4rI0gDhI0Tk\nhnrIVuLsc0Tkjob+xhD5/1RETo90vj75Z4rIz33O7xGR68OkCWTmFCr+Cuc/ucc5f1FEPnbC/xIk\nTSsRyXfirBaRV0WknXPtv5ywVSKyVEQGOoaxv/VJf4GIvBjpvPxkvMYx+1rlPHc/8Pm9XQOl8SVu\nFhWhUNU5wBwRGYYxFr0FjBJivtrfIiJtMAvHvOIyzyUNlOUTp5xI81OgArNC71FEJFVVG2QJ4kcm\n8HPMNzW3TALyG1nur1T1gxDXHwU2q+okABHpCzQXkWswC67+UFUPO8rRLUxZkczLl9dUdaGT5++B\nGwDXlh8JqVQuycB8MQ/EySLyCtAdmKeqc8RMFTlNVe8TkRUYRemDsTi4XFUPichszJ+xBWPhjK9i\nO2+ow0AWxn7sJ6q6W0SmAtcCxUAv4Cr1sXlz8ngI+B5j0DobGAH0F5ESVb1KRP6JMQY+XUT+A/ij\nI38axntlUX3KB6YCA53fOtsR5ccicjXGyuAaVd3iI+O1QGcn/nvAY5i1Hk8BajGLqZb43eefY14M\nm33CHhWRQ8ATqnrcC0/MiscXYywaAFDVTc618cB/q+phJ3wPsEdEziEAkczLHzUG215aA5/5/N4y\n4K5Q6ZNRqa4QkdXAACDYXJwuGMvjg8B6MSvi+rNCVW8XkWeAS0Tka6Cvqp7nWE4Hayp9pqoTRGQG\ncLWIvIp5k+UCrYAvA6QZDfxWVd8VkRRVrRWRJcBzPm/1ThiLiK9EZCLGROsWETkVeA04v57lPwr0\nUdWLAUQkFyhXs9T2tZiVrqZ5BVTVl0Xk96o6zIl/O7BJVX8vIhdhXgqjfX+Uqnpr2UpnP01VK0Tk\nZOB9EVmvqr73owNQoYEtDroA/woQHoxI5lUHMfO8bgeqgAfhuN9bFiwdxH+gogozb8pLCycsFG+p\n6oWYt3GwCXdbVHW/86baTOCq37v871eYt34vYD2AU8t8EyRv/3TdME2QGlX9DlPL+TMbY0H+EsbQ\nNBBf+/xpfYGxTq3xCsboszHlB0sbimyMmQ/OvneY+OgxQ9m9mNquv4iM8faJgXKgvYgEWibsX5ga\n1J9gz0jE8hKRHnJs0mQP5zc8r6p9gUVAvfrU8VaqYqCfHJvrMhxjNxcWVf0bcIaInB3gcm8RaSMi\nzYBzgO2BsvA5FmAbxrsJzgDCqcGK9ktXCpwtIs1EJAPzMPqzR1V/ian9pjsd32qObyn49qM+A/6s\nqsOcmsN35qrb8v3zD5TWnxqnWQXmv8lzjvOc85CI8QmNiKRjatatqrrI+R0j1cyxWgr8t0+as0Uk\nC7MO5CwxC7AiIieLyEDMBMvzfeQaDmyIcF4l3nutqiVy/EzgfRyriV0R1+afGh/CDwDLnXb4XoI3\n6QLxODAd0/zxpRR4FugJzHX6HeFk2SBmWvpaTO3mygeXqn4jIi9jVuXdilkctNov2lQRuRTzEntP\nVb9z3ty/F5HPVfVWv/jPAn8QkeXO+d8J8rYMUf4uzFv4r8BTbn4L5q38loj8zZHhzyKyCqOME1yk\nf8UZQEoD5qvqZwHiTAX+n4isxNyPMoyV+EJnQGGZmIUvqzGOKz4XkTeB1SJSg2ne3hOFvHy5Q0R+\n5BzvxRjnusYa1EYAEUlzRpl+gLHA7hWhEbykKN9yPMk4UJGITHfebCdh1j6I9QMd7/ItPtiaymKJ\nMPEeqLBYTjisUlksEcYqlcUSYaxSWSwRxiqVxRJh/j9x8Y/YU4JifgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x123dd4828>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(3,2))\n",
    "ax = fig.add_subplot(111)\n",
    "ax.hist(strongest_fake, 60, normed=True, alpha=0.7, color='gray', label='Shuffled UTRs')\n",
    "ax.hist(strongest_real, 60, normed=True, alpha=0.7, label='Real UTRs')\n",
    "ax.axvline(-4.5, c='r')\n",
    "ax.axvline(-3.5, c='r', linestyle='--')\n",
    "ax.axvline(-5.5, c='r', linestyle=':')\n",
    "ax.set_xlabel('UTR binding strength to \"5-CCUCCU-3\"')\n",
    "ax.set_ylabel('Probability density')\n",
    "fig.subplots_adjust(top=0.8, left=0.2, right=0.9, bottom=0.25)  # create some space below the plots by increasing the bottom-value\n",
    "ax.legend(loc='upper center', ncol=2, bbox_to_anchor=(0.15, 0.9, 0.7, 0.4), fontsize=8)\n",
    "plt.savefig('../Results/Figures/SD_threshold.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Actually dig into data to test for conservation of SD motifs\n",
    "The below analysis is limited to ONLY 4-fold redundant codon sites that are within an aSD sequence binding site. Critically, for each one that I find, I compare the conservation at that site to the same synonymous codon or nucleotide context somewhere else in the same gene. With these paired lists, we can accurately compare whether there is any conservation differences while ruling out gene-specific effects."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Positive control data. Find genes that have another gene directly 3' of them which are thus likely to contain a real SD sequence in their 3' terminus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "391 1003 1394\n"
     ]
    }
   ],
   "source": [
    "preceding_genes = []\n",
    "starts_positive_strand = []\n",
    "starts_negative_strand = []\n",
    "genome = list(SeqIO.parse('/Users/adamhockenberry/Projects/2017/Genome_database/Data/GBFs/511145.12.PATRIC.gbf', 'genbank'))\n",
    "for record in genome:\n",
    "    for feature in record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            if feature.location.strand == 1:\n",
    "                starts_positive_strand.append(feature.location.start)\n",
    "            elif feature.location.strand == -1:\n",
    "                starts_negative_strand.append(feature.location.end)\n",
    "for i,j in enumerate(starts_positive_strand):\n",
    "    starts_positive_strand[i] = list(range(j-20,j+20,1))\n",
    "for i,j in enumerate(starts_negative_strand):\n",
    "    starts_negative_strand[i] = list(range(j-20,j+20,1))\n",
    "for record in genome:\n",
    "    for feature in record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            if feature.location.strand == 1:\n",
    "                end = feature.location.end\n",
    "                for i in starts_positive_strand:\n",
    "                    if end in i:\n",
    "                        preceding_genes.append(feature.qualifiers['locus_tag'][0])\n",
    "            elif feature.location.strand == -1:\n",
    "                end = feature.location.start\n",
    "                for i in starts_negative_strand:\n",
    "                    if end in i:\n",
    "                        preceding_genes.append(feature.qualifiers['locus_tag'][0])\n",
    "                        \n",
    "preceding_genes = [i for i in seq_dicty[base_genome].keys() if i in preceding_genes]\n",
    "not_preceding_genes = [i for i in seq_dicty[base_genome].keys() if i not in preceding_genes]\n",
    "print(len(preceding_genes), len(not_preceding_genes), len(seq_dicty[base_genome].keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "three_prime_limitation_dict = {}\n",
    "for gene_name, nts in seq_dicty[base_genome].items():\n",
    "    three_prime_limitation_dict[gene_name] = list(range(len(nts)-50, len(nts)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing conservation of a particular 4-fold redundant position that is within a strong binding sequence to a randomly chosen identical 4-fold redundant codon elsewhere in the sequence\n",
    "Note I'm currently looking at a 6nt aSD sequence and only considering 4-fold redundant codons in positions 2,3,4, or 5. I'm also taking care not to double count any one position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_sd_sites(nt_sequence, asd_seq, threshold, fragment_energy_dict, position_limitation_list, upper_threshold=-20, min_position=100, max_position=-50):\n",
    "    codon_list = [nt_sequence[nt:nt+3] for nt in range(0, len(nt_sequence), 3)]\n",
    "    energy_list = []\n",
    "    for i in range(0, len(nt_sequence)-len(asd_seq)):\n",
    "        fragment = nt_sequence[i:i+len(asd_seq)]\n",
    "        if fragment.count('-') == 0:\n",
    "            energy = fragment_energy_dict[fragment.replace('T', 'U')]\n",
    "            energy_list.append(energy)\n",
    "        else:\n",
    "            energy_list.append(np.nan)\n",
    "    #######This gets all positions below the given threshold\n",
    "    sd_positions_all_nts = [list(range(i, i+len(asd_seq))) for i,j in enumerate(energy_list) if j < threshold]\n",
    "    sd_positions_all_nts = [item for sublist in sd_positions_all_nts for item in sublist]\n",
    "    sd_positions_all_nts = list(set(sd_positions_all_nts))\n",
    "    \n",
    "    #######This gets all \"start\" position within the threshold range. Subtle but important difference\n",
    "    if position_limitation_list != False:\n",
    "        sd_positions_start = np.array([i for i,j in enumerate(energy_list) if upper_threshold < j and j < threshold and i in position_limitation_list])\n",
    "    else:\n",
    "        sd_positions_start = np.array([i for i,j in enumerate(energy_list) if upper_threshold < j and j < threshold and i > min_position and i < len(nt_sequence)+max_position])\n",
    "    \n",
    "    np.random.shuffle(sd_positions_start) #Shuffle this so that I'm not always analyzing the 5' ones every single time later on\n",
    "    return sd_positions_all_nts, sd_positions_start, energy_list, codon_list\n",
    "\n",
    "def get_control_sites_codon(nt_sequence, codon_list, sd_positions_all_nts, analyzed_positions, codon, min_position=100, max_position=-50):\n",
    "    control_sites = [site*3+2 for site, x in enumerate(codon_list) if x == codon]\n",
    "    control_sites = [site for site in control_sites if site > min_position and site < len(nt_sequence)+max_position]\n",
    "    control_sites = [site for site in control_sites if site not in sd_positions_all_nts]\n",
    "    control_sites = [site for site in control_sites if site not in analyzed_positions]\n",
    "    return control_sites\n",
    "\n",
    "def get_control_sites_tri_nt(nt_sequence, codon_list, sd_positions_all_nts, analyzed_positions, tri_nt, redundant_codons, min_position=100, max_position=-50):\n",
    "    control_sites = [m.start()+1 for m in re.finditer(tri_nt, nt_sequence)]\n",
    "    control_sites = [site for site in control_sites if site > min_position and site < len(nt_sequence)+max_position]\n",
    "    control_sites = [site for site in control_sites if site%3 == 2]\n",
    "    control_sites = [site for site in control_sites if nt_sequence[site-2:site+1] in redundant_codons]\n",
    "    control_sites = [site for site in control_sites if site not in sd_positions_all_nts]\n",
    "    control_sites = [site for site in control_sites if site not in analyzed_positions]\n",
    "    return control_sites\n",
    "\n",
    "def get_possible_energies(fragment, codon, all_codons, modifier, fragment_energy_dict):\n",
    "    possible_energies = []\n",
    "    for possible_codon in all_codons:\n",
    "        if possible_codon != codon:\n",
    "            if modifier == 2:\n",
    "                possible_fragment = possible_codon + fragment[-3:]\n",
    "            elif modifier == 3:\n",
    "                possible_fragment = fragment[0] + possible_codon + fragment[-2:]\n",
    "            elif modifier == 1:\n",
    "                possible_fragment = possible_codon[-2:] + fragment[-4:]\n",
    "            elif modifier == 4:\n",
    "                possible_fragment = fragment[0:2] + possible_codon + fragment[-1]\n",
    "            assert len(possible_fragment) == 6\n",
    "            possible_energies.append(fragment_energy_dict[possible_fragment.replace('T', 'U')])\n",
    "    return possible_energies\n",
    "\n",
    "\n",
    "def paired_nt_analysis(dicty_to_analyze,\\\n",
    "                      genome_seq_dicty,\\\n",
    "                      fragment_energy_dict,\\\n",
    "                      asd_seq,\\\n",
    "                      conservation_dicty,\\\n",
    "                      nt_conservation_dicty,\\\n",
    "                      redundant_codons=redundant_codons,\\\n",
    "                      threshold=-4.5,\\\n",
    "                      upper_threshold=-20,\\\n",
    "                      min_position=100,\\\n",
    "                      max_position=-50,\\\n",
    "                      subset_names=False,\\\n",
    "                      only_conserved_aa=False,\\\n",
    "                      position_limitation_dict=False,\\\n",
    "                      internal_control='',\\\n",
    "                      nt_strength_restriction=''):\n",
    "    '''\n",
    "    This is my rather monsterous main function that has a lot of flags so be careful! \n",
    "    However, its complexity provides the benefit that I know all analyses are run through\n",
    "    a common framework. For the most part, all of these flags are ultimately limiting the \n",
    "    data according to various criteria at different points in the pipeline.\n",
    "    '''\n",
    "    sds = []\n",
    "    others = []\n",
    "    analyzed_sd_dict = {}\n",
    "    analyzed_control_dict = {}\n",
    "    \n",
    "    all_sds = []\n",
    "    all_others = []\n",
    "        \n",
    "    for gene_name, seq in list(genome_seq_dicty.items())[:]:\n",
    "\n",
    "        analyzed_sds = []\n",
    "        analyzed_controls = []\n",
    "        \n",
    "        gene_sds = []\n",
    "        gene_others = []\n",
    "        \n",
    "        ###########Skip ahead if this gene is not in my dictionary to analyze\n",
    "        if gene_name not in dicty_to_analyze.keys():\n",
    "            continue\n",
    "        ###########Skip ahead if I want to only analyze a subset and this gene isn't in it\n",
    "        if subset_names:\n",
    "            if gene_name not in subset_names:\n",
    "                continue\n",
    "        ###########Hang on to this information for now to eventually limit where in the gene to consider, if relevant\n",
    "        position_limitation_list = False\n",
    "        if position_limitation_dict != False:\n",
    "            position_limitation_list = position_limitation_dict[gene_name]\n",
    "            \n",
    "        ##################################################################################################\n",
    "        ###GET LOCATIONS OF ALL SD-LIKE SEQUENCES THAT I MIGHT WANT TO ANALYZE FOR THIS GENE       \n",
    "        sd_positions_all_nts, sd_positions_start, energy_list, codon_list =\\\n",
    "                        get_sd_sites(seq, asd_seq, threshold, fragment_energy_dict,\\\n",
    "                                     position_limitation_list=position_limitation_list,\\\n",
    "                                     upper_threshold=upper_threshold, min_position=min_position,\\\n",
    "                                     max_position=max_position)\n",
    "\n",
    "        for i in sd_positions_start:\n",
    "            energy = energy_list[i]\n",
    "            modifiers = []\n",
    "            if i%3 == 0:\n",
    "                modifiers = [2]\n",
    "            elif i%3 == 2:\n",
    "                modifiers = [3]\n",
    "            elif i%3 == 1:\n",
    "                modifiers = [1,4]\n",
    "            for modifier in modifiers:\n",
    "                ##################################################################################################\n",
    "                #If I'm only considering particular regions, make sure that this location is one of them\n",
    "                if position_limitation_dict != False:\n",
    "                    if i+modifier not in position_limitation_dict[gene_name]:\n",
    "                        continue\n",
    "\n",
    "                analyzed_positions = analyzed_sds + analyzed_controls\n",
    "                \n",
    "                ##################################################################################################\n",
    "                #Make sure that the location is conserved, and if TRINT that the next position is conserved\n",
    "                if only_conserved_aa == True:\n",
    "                    if conservation_dicty[gene_name][int(np.floor((i+modifier)/3.))] != 'C':\n",
    "                        continue\n",
    "                if only_conserved_aa == True and internal_control=='TRINT':\n",
    "                    if nt_conservation_dicty[gene_name][i+modifier+1] != 'C':\n",
    "                        continue\n",
    "                \n",
    "                codon = seq[i+modifier-2:i+modifier+1]\n",
    "                tri_nt = seq[i+modifier-1:i+modifier+2]\n",
    "                if codon in redundant_codons:\n",
    "                    ##################################################################################################\n",
    "                    ###RUN A FILTER ACCORDING TO WHETHER I ONLY WANT TO SELECT STRONGEST OR WEAKEST POSSIBLE SEQUENCES\n",
    "                    if nt_strength_restriction == 'None':\n",
    "                        pass\n",
    "                    else:\n",
    "                        fragment = genome_seq_dicty[gene_name][i:i+len(asd_seq)]\n",
    "                        all_codons = aa_to_codon_dict[codon_to_aa_dict[codon]]\n",
    "                        all_codons = [temp for temp in all_codons if temp[:2] == codon[:2]]###Dealing with 6-fold redundant codons\n",
    "                        possible_energies = get_possible_energies(fragment, codon, all_codons, modifier, fragment_energy_dict)\n",
    "                        possible_energies = sorted(possible_energies)\n",
    "                        assert len(possible_energies) == 3, possible_energies\n",
    "                        if energy in possible_energies:###Because I don't know how to treat these cases\n",
    "                            continue\n",
    "                        insertion_point = bisect(possible_energies, energy)\n",
    "                        #################################################################\n",
    "                        ###Below will get rid of the ones that aren't included in my flag\n",
    "                        if int(nt_strength_restriction) != insertion_point:\n",
    "                            continue \n",
    "                            \n",
    "                    ##################################################################################################\n",
    "                    ###GET CONTROL SITES ACCORDING TO EITHER OF TWO DIFFERENT NULL MODELS\n",
    "                    if internal_control == 'CODON':\n",
    "                        control_sites =\\\n",
    "                            get_control_sites_codon(seq, codon_list, sd_positions_all_nts, analyzed_positions, codon,\\\n",
    "                                                   min_position=min_position, max_position=max_position)\n",
    "                    elif internal_control == 'TRINT':\n",
    "                        control_sites =\\\n",
    "                            get_control_sites_tri_nt(seq, codon_list, sd_positions_all_nts, analyzed_positions, tri_nt, redundant_codons,\\\n",
    "                                                    min_position=min_position, max_position=max_position)\n",
    "                    else:\n",
    "                        print(\"YOU DID NOT PROPERLY SPECIFY A CONTROL METHOD\")\n",
    "                        break\n",
    "                    ##################################################################################################\n",
    "                    ###Limit the controls based on conservation of the amino acid\n",
    "                    if only_conserved_aa == True and internal_control == 'CODON':\n",
    "                        control_sites = [site for site in control_sites if conservation_dicty[gene_name][int(np.floor(site/3.))] == 'C']\n",
    "                    elif only_conserved_aa == True and internal_control == 'TRINT':\n",
    "                        control_sites = [site for site in control_sites if conservation_dicty[gene_name][int(np.floor(site/3.))] == 'C']\n",
    "                        control_sites = [site for site in control_sites if nt_conservation_dicty[gene_name][site+1] == 'C']\n",
    "\n",
    "                    ##################################################################################################\n",
    "                    ###IF ALL LOOKS WELL CHOOSE A CONTROL AND GO FORWARD WITH ANALYSIS\n",
    "                    gene_sds.append(i+modifier)\n",
    "                    gene_others.extend(control_sites)\n",
    "                    if len(control_sites) > 0 and i+modifier not in analyzed_positions:\n",
    "                        choice = random.choice(control_sites)\n",
    "                        sds.append(dicty_to_analyze[gene_name][i+modifier])\n",
    "                        others.append(dicty_to_analyze[gene_name][choice])\n",
    "                        analyzed_sds.append(i+modifier)\n",
    "                        analyzed_controls.append(choice)\n",
    "                        #######################################################################\n",
    "                        ###Fail safe to make sure I chose the right codons and context controls\n",
    "                        if internal_control == 'CODON':\n",
    "                            assert seq_dicty[base_genome][gene_name][i+modifier-2:i+modifier+1] == seq_dicty[base_genome][gene_name][choice-2:choice+1]\n",
    "                        if internal_control == 'TRINT':\n",
    "                            assert seq_dicty[base_genome][gene_name][i+modifier-1:i+modifier+2] == seq_dicty[base_genome][gene_name][choice-1:choice+2]\n",
    "        gene_sds = list(set(gene_sds))\n",
    "        gene_others = list(set(gene_others))\n",
    "        \n",
    "        all_sds.extend([dicty_to_analyze[gene_name][position] for position in gene_sds])\n",
    "        all_others.extend([dicty_to_analyze[gene_name][position] for position in gene_others])\n",
    "        analyzed_sd_dict[gene_name] = analyzed_sds\n",
    "        analyzed_control_dict[gene_name] = analyzed_controls\n",
    "    return sds, others, analyzed_sd_dict, analyzed_control_dict, all_sds, all_others"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main function\n",
    "There are a lot of flags but the analysis below is essentially the analysis that is used throughout the manuscript. Thresholds, null models, etc can all be altered below to accomplish various analyses that were depicted in the manuscript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#####0\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.9728406158325256, Ratio:0.3439312830646899\n",
      "Coin flip expt: 0.1935483870967742\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.1022163826029683, Ratio:0.29204380324097734\n",
      "Coin flip expt: 0.13333333333333333\n",
      "#####1\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.7008901069112174, Ratio:0.3785614545122672\n",
      "Coin flip expt: 0.25806451612903225\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.2745313721685383, Ratio:0.2766756423688975\n",
      "Coin flip expt: 0.13333333333333333\n",
      "#####2\n",
      "Lengths:31, 31, Means:1.0224528873501033, 3.2360298388181548, Ratio:0.3159590418744464\n",
      "Coin flip expt: 0.22580645161290322\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.238801893780632, Ratio:0.2797278439881019\n",
      "Coin flip expt: 0.2\n",
      "#####3\n",
      "Lengths:31, 31, Means:1.0224528873501033, 3.071263251123932, Ratio:0.3329095566705119\n",
      "Coin flip expt: 0.12903225806451613\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.401088073683417, Ratio:0.26638036158547573\n",
      "Coin flip expt: 0.06666666666666667\n",
      "#####4\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.9768867330380018, Ratio:0.3434638194334857\n",
      "Coin flip expt: 0.22580645161290322\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.1927028473856662, Ratio:0.2837668001560806\n",
      "Coin flip expt: 0.2\n",
      "#####5\n",
      "Lengths:31, 31, Means:1.0224528873501033, 3.1173384146869316, Ratio:0.3279890571177484\n",
      "Coin flip expt: 0.16129032258064516\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.223169192038378, Ratio:0.2810845527717647\n",
      "Coin flip expt: 0.2\n",
      "#####6\n",
      "Lengths:31, 31, Means:1.0224528873501033, 3.160817834244468, Ratio:0.32347732168326643\n",
      "Coin flip expt: 0.12903225806451613\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.6258608292717183, Ratio:0.2498670284137219\n",
      "Coin flip expt: 0.06666666666666667\n",
      "#####7\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.8346458003677575, Ratio:0.36069864080283104\n",
      "Coin flip expt: 0.22580645161290322\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.271565617156126, Ratio:0.2769264556702921\n",
      "Coin flip expt: 0.13333333333333333\n",
      "#####8\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.7669758337789676, Ratio:0.36951999177878586\n",
      "Coin flip expt: 0.25806451612903225\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.029854917927459, Ratio:0.29901863138435886\n",
      "Coin flip expt: 0.2\n",
      "#####9\n",
      "Lengths:31, 31, Means:1.0224528873501033, 3.4039728365512176, Ratio:0.3003704601785294\n",
      "Coin flip expt: 0.22580645161290322\n",
      "Lengths:15, 15, Means:0.9059830708518377, 2.997559487879002, Ratio:0.30224023059935623\n",
      "Coin flip expt: 0.13333333333333333\n",
      "#####10\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.687791925127694, Ratio:0.3804062650056246\n",
      "Coin flip expt: 0.2903225806451613\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.4497390379657187, Ratio:0.2626236538129818\n",
      "Coin flip expt: 0.06666666666666667\n",
      "#####11\n",
      "Lengths:31, 31, Means:1.0224528873501033, 3.102660901818837, Ratio:0.329540649044413\n",
      "Coin flip expt: 0.1935483870967742\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.183367098031822, Ratio:0.28459899312648523\n",
      "Coin flip expt: 0.26666666666666666\n",
      "#####12\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.9985625861431098, Ratio:0.3409810060577157\n",
      "Coin flip expt: 0.22580645161290322\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.232761602065328, Ratio:0.28025050479225827\n",
      "Coin flip expt: 0.2\n",
      "#####13\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.6420542369181574, Ratio:0.3869916344119985\n",
      "Coin flip expt: 0.2903225806451613\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.345832572916377, Ratio:0.27077955967836803\n",
      "Coin flip expt: 0.13333333333333333\n",
      "#####14\n",
      "Lengths:31, 31, Means:1.0224528873501033, 3.0262824078929906, Ratio:0.33785772427695293\n",
      "Coin flip expt: 0.22580645161290322\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.2580145433926315, Ratio:0.2780782770565599\n",
      "Coin flip expt: 0.2\n",
      "#####15\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.7743074933980933, Ratio:0.368543461668612\n",
      "Coin flip expt: 0.1935483870967742\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.52978473162281, Ratio:0.25666808027562465\n",
      "Coin flip expt: 0.2\n",
      "#####16\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.652374107273223, Ratio:0.3854859254380361\n",
      "Coin flip expt: 0.2903225806451613\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.346740941080788, Ratio:0.27070606503509703\n",
      "Coin flip expt: 0.2\n",
      "#####17\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.915573861842941, Ratio:0.3506866695202873\n",
      "Coin flip expt: 0.3225806451612903\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.1851900807108877, Ratio:0.2844361083309777\n",
      "Coin flip expt: 0.06666666666666667\n",
      "#####18\n",
      "Lengths:31, 31, Means:1.0224528873501033, 3.028369255137518, Ratio:0.33762490674330725\n",
      "Coin flip expt: 0.22580645161290322\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.2980764314731332, Ratio:0.2747004472686424\n",
      "Coin flip expt: 0.2\n",
      "#####19\n",
      "Lengths:31, 31, Means:1.0224528873501033, 2.8037676405591316, Ratio:0.36467104925506744\n",
      "Coin flip expt: 0.25806451612903225\n",
      "Lengths:15, 15, Means:0.9059830708518377, 3.328096501178715, Ratio:0.27222259646947283\n",
      "Coin flip expt: 0.2\n"
     ]
    }
   ],
   "source": [
    "dicty_to_analyze = dicty_hyphy #####If this changes, make sure to manually change file_name_modifiers\n",
    "threshold=-4.5\n",
    "upper_threshold=-20.\n",
    "subset_names=preceding_genes\n",
    "# subset_names=False\n",
    "only_conserved_aa=True\n",
    "position_limitation_dict=three_prime_limitation_dict\n",
    "# position_limitation_dict=False\n",
    "controls=('CODON', 'TRINT')\n",
    "nt_strength_restriction='None'\n",
    "\n",
    "\n",
    "mean_randomizations = 20\n",
    "stat_randomizations = 20\n",
    "\n",
    "#########################################################\n",
    "####Putative aSD sequence to use in considering hybridization energies\n",
    "with open('../Data/json-energyRef-CCUCCU.txt') as infile:\n",
    "    fragment_energy_dict = json.load(infile)\n",
    "asd_seq = 'CCUCCU'\n",
    "#########################################################\n",
    "#########################################################\n",
    "#\n",
    "#\n",
    "#\n",
    "#########################################################\n",
    "####For writing files later\n",
    "file_name_modifiers = '{}_{}_Threshold={}to{}_Conserved={}_NtRestriction={}'.format('hyphy', species_name, upper_threshold, threshold, only_conserved_aa, nt_strength_restriction)\n",
    "if subset_names != False or position_limitation_dict != False:\n",
    "    file_name_modifiers = file_name_modifiers + '_Subset=preceding_Position_limitation=3prime'\n",
    "#########################################################\n",
    "#########################################################\n",
    "#\n",
    "#\n",
    "#########################################################\n",
    "#These containers are for keeping track of randomizations to assess significance\n",
    "actual_ratios_codon = []\n",
    "shuffled_ratios_codon = {}\n",
    "actual_ratios_trint = []\n",
    "shuffled_ratios_trint = {}\n",
    "\n",
    "for i in range(mean_randomizations):\n",
    "    shuffled_ratios_codon[i] = []\n",
    "    shuffled_ratios_trint[i] = []\n",
    "    print('#####{}'.format(i))\n",
    "    sds_codon, others_codon, analyzed_sd_dict, analyzed_ctrl_dict, all_sds_codon, all_others_codon = paired_nt_analysis(dicty_to_analyze,\\\n",
    "                                     seq_dicty[base_genome],\\\n",
    "                                     fragment_energy_dict,\\\n",
    "                                     asd_seq,\\\n",
    "                                     conservation_dicty,\\\n",
    "                                     nt_conservation_dicty,\\\n",
    "                                     threshold=threshold,\\\n",
    "                                     upper_threshold=upper_threshold,\\\n",
    "                                     subset_names=subset_names,\\\n",
    "                                     only_conserved_aa=only_conserved_aa,\\\n",
    "                                     position_limitation_dict=position_limitation_dict,\\\n",
    "                                     internal_control=controls[0],\\\n",
    "                                     nt_strength_restriction=nt_strength_restriction)\n",
    "    actual_ratios_codon.append(np.mean(sds_codon)/np.mean(others_codon))\n",
    "    print('Lengths:{}, {}, Means:{}, {}, Ratio:{}'.format(len(sds_codon), len(others_codon),\\\n",
    "                                                          np.mean(sds_codon), np.mean(others_codon),\\\n",
    "                                                          np.mean(sds_codon)/np.mean(others_codon)))\n",
    "    diffs = np.array(sds_codon)-np.array(others_codon)\n",
    "    print('Coin flip expt:', len([i for i in diffs if i > 0])/len([i for i in diffs if i != 0]))\n",
    "    for j in range(stat_randomizations):\n",
    "        combined = sds_codon + others_codon\n",
    "        random.shuffle(combined)\n",
    "        random_sds = combined[:len(sds_codon)]\n",
    "        random_others = combined[len(sds_codon):]\n",
    "        shuffled_ratios_codon[i].append(np.mean(random_sds)/np.mean(random_others))\n",
    "        \n",
    "        \n",
    "    sds_trint, others_trint, analyzed_sd_dict, analyzed_ctrl_dict, all_sds_trint, all_others_trint = paired_nt_analysis(dicty_to_analyze,\\\n",
    "                                     seq_dicty[base_genome],\\\n",
    "                                     fragment_energy_dict,\\\n",
    "                                     asd_seq,\\\n",
    "                                     conservation_dicty,\\\n",
    "                                     nt_conservation_dicty,\\\n",
    "                                     threshold=threshold,\\\n",
    "                                     upper_threshold=upper_threshold,\\\n",
    "                                     subset_names=subset_names,\\\n",
    "                                     only_conserved_aa=only_conserved_aa,\\\n",
    "                                     position_limitation_dict=position_limitation_dict,\\\n",
    "                                     internal_control=controls[1],\\\n",
    "                                     nt_strength_restriction=nt_strength_restriction)\n",
    "    actual_ratios_trint.append(np.mean(sds_trint)/np.mean(others_trint))\n",
    "    print('Lengths:{}, {}, Means:{}, {}, Ratio:{}'.format(len(sds_trint), len(others_trint),\\\n",
    "                                                          np.mean(sds_trint), np.mean(others_trint),\\\n",
    "                                                          np.mean(sds_trint)/np.mean(others_trint)))\n",
    "    diffs = np.array(sds_trint)-np.array(others_trint)\n",
    "    print('Coin flip expt:', len([i for i in diffs if i > 0])/len([i for i in diffs if i != 0]))\n",
    "    for j in range(stat_randomizations):\n",
    "        combined = sds_trint + others_trint\n",
    "        random.shuffle(combined)\n",
    "        random_sds = combined[:len(sds_trint)]\n",
    "        random_others = combined[len(sds_trint):]\n",
    "        shuffled_ratios_trint[i].append(np.mean(random_sds)/np.mean(random_others))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dfA = pd.DataFrame({'Actual_rats_codon':actual_ratios_codon})\n",
    "dfB = pd.DataFrame({'Shuffled_rats_codon':[j[0] for i,j in shuffled_ratios_codon.items()]})\n",
    "dfC = pd.DataFrame({'Actual_rats_trint':actual_ratios_trint})\n",
    "dfD = pd.DataFrame({'Shuffled_rats_trint':[j[0] for i,j in shuffled_ratios_trint.items()]})\n",
    "df = pd.concat([dfA[['Actual_rats_codon']],dfB[['Shuffled_rats_codon']],\\\n",
    "               dfC[['Actual_rats_trint']],dfD[['Shuffled_rats_trint']]], axis=1)\n",
    "df['N_(codon)'] = len(sds_codon)\n",
    "df['N_(trint)'] = len(sds_trint)\n",
    "df.to_csv('../Results/Dataframes/df_{}_ratios.csv'.format(file_name_modifiers), index=False)\n",
    "\n",
    "\n",
    "dfA = pd.DataFrame({'SD-like(codon)':all_sds_codon})\n",
    "dfB = pd.DataFrame({'SD-like(trint)':all_sds_trint})\n",
    "dfC = pd.DataFrame({'Control\\n(codon)':all_others_codon})\n",
    "dfD = pd.DataFrame({'Control\\n(context)':all_others_trint})\n",
    "df = pd.concat([dfA[['SD-like(codon)']],dfB[['SD-like(trint)']], dfC[['Control\\n(codon)']], dfD[['Control\\n(context)']]], axis=1)\n",
    "df.to_csv('../Results/Dataframes/df_{}_all.csv'.format(file_name_modifiers), index=False)\n",
    "\n",
    "dist_of_codon_means = []\n",
    "for j in range(stat_randomizations):\n",
    "    temp = []\n",
    "    for i in range(mean_randomizations):\n",
    "        temp.append(shuffled_ratios_codon[i][j])\n",
    "    dist_of_codon_means.append(np.mean(temp))\n",
    "\n",
    "dist_of_trint_means = []\n",
    "for j in range(stat_randomizations):\n",
    "    temp = []\n",
    "    for i in range(mean_randomizations):\n",
    "        temp.append(shuffled_ratios_trint[i][j])\n",
    "    dist_of_trint_means.append(np.mean(temp))\n",
    "    \n",
    "dfA = pd.DataFrame({'Names':['Actual']+list(range(len(dist_of_codon_means)))})\n",
    "dfB = pd.DataFrame({'Values(codon)':[np.mean(actual_ratios_codon)]+dist_of_codon_means})\n",
    "dfC = pd.DataFrame({'Values(trint)':[np.mean(actual_ratios_trint)]+dist_of_trint_means})\n",
    "df = pd.concat([dfA[['Names']],dfB[['Values(codon)']], dfC[['Values(trint)']]], axis=1)\n",
    "df.to_csv('../Results/Dataframes/df_{}_stat_sig.csv'.format(file_name_modifiers), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###Saved this data to make prettier figures but because I'm ants-y and want to see how they look\n",
    "plt.figure()\n",
    "plt.hist(dist_of_trint_means)\n",
    "plt.axvline(np.mean(actual_ratios_trint), c='r')\n",
    "\n",
    "plt.figure()\n",
    "plt.hist(dist_of_codon_means)\n",
    "plt.axvline(np.mean(actual_ratios_codon), c='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Put rates into context\n",
    "Similar paired analysis of 2-fold vs 4-fold redundant codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "genome_seq_dicty = seq_dicty[base_genome]\n",
    "dicty_to_analyze = dicty_hyphy\n",
    "\n",
    "ratios_all = []\n",
    "for randomization in range(3):\n",
    "    two_fold = []\n",
    "    four_fold = []\n",
    "    two_fold_all = []\n",
    "    four_fold_all = []\n",
    "    for gene_name, nt_sequence in list(genome_seq_dicty.items())[:]:\n",
    "        if gene_name not in dicty_to_analyze.keys():\n",
    "            continue\n",
    "        two_fold_gene = []\n",
    "        four_fold_gene = []\n",
    "        codon_list = [nt_sequence[nt:nt+3] for nt in range(0, len(nt_sequence), 3)]\n",
    "        for i, codon in enumerate(codon_list):\n",
    "            if codon == '---':\n",
    "                continue\n",
    "            if i < 33:\n",
    "                continue\n",
    "            if i > len(codon_list)-16:\n",
    "                continue\n",
    "            if conservation_dicty[gene_name][i] != 'C':\n",
    "                continue\n",
    "            if codon in redundant_codons:\n",
    "                four_fold_gene.append(dicty_to_analyze[gene_name][i*3+2])\n",
    "            elif codon_to_aa_dict[codon] in ['F', 'D', 'E', 'K', 'N', 'Q', 'H', 'Y', 'C']:\n",
    "                two_fold_gene.append(dicty_to_analyze[gene_name][i*3+2])\n",
    "        smallest_set = min([len(two_fold_gene), len(four_fold_gene)])\n",
    "        two_fold_truncated = random.sample(two_fold_gene, smallest_set)\n",
    "        four_fold_truncated = random.sample(four_fold_gene, smallest_set)\n",
    "        two_fold.extend(two_fold_truncated)\n",
    "        four_fold.extend(four_fold_truncated)\n",
    "        two_fold_all.extend(two_fold_gene)\n",
    "        four_fold_all.extend(four_fold_gene)\n",
    "    ratio_i = np.mean(four_fold)/np.mean(two_fold)\n",
    "    ratios_all.append(ratio_i)\n",
    "    print(randomization, ratio_i, len(four_fold), len(two_fold), len(two_fold_all), len(four_fold_all))\n",
    "print(np.mean(four_fold_all)/np.mean(two_fold_all))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.boxplot([two_fold, four_fold]);\n",
    "plt.ylim(0,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Protein Domain analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv('../Data/foldingFromPNAS_v2.csv')\n",
    "df['bNumber'] = ''\n",
    "df['patricNumber'] = ''\n",
    "print(len(df.index))\n",
    "\n",
    "for index in df.index:\n",
    "    names = df.loc[index]['Gene names'].split(' ')\n",
    "    for name in names:\n",
    "        if name in reference_name_dict.keys():\n",
    "            df.set_value(index, 'bNumber', name)\n",
    "            try:\n",
    "                df.set_value(index, 'patricNumber', reference_name_dict[name])\n",
    "            except KeyError:\n",
    "                pass\n",
    "\n",
    "df = df[df['patricNumber'] != '']\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Choose whether to stratify the data according to Lm\n",
    "A lot of degrees of potential degrees of freedom here which can be dangerous. Be rational."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df = df[df['Lm (residues)'].isnull() == False]\n",
    "# df = df[(df['Lm (residues)'] > 0) & (df['Lm (residues)'] <= 50)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `domain_dict` will specify what regions in what proteins I want to actually consider "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df['Domain_beg_nt'] = int()\n",
    "df['Domain_end_nt'] = int()\n",
    "for index in df.index:\n",
    "    domain_beg, domain_end = df.loc[index]['Codon range of domain'].split('-')\n",
    "    df.set_value(index, 'Domain_beg_nt', int(domain_beg)*3)\n",
    "    df.set_value(index, 'Domain_end_nt', int(domain_end)*3)\n",
    "\n",
    "\n",
    "df['Domain_beg_in_patric'] = int()\n",
    "df['Domain_end_in_patric'] = int()\n",
    "for index in df.index:    \n",
    "    ref_seq_beg = ref_genome_seq_ref_dict[df.loc[index]['bNumber']][int(df.loc[index]['Domain_beg_nt']):int(df.loc[index]['Domain_beg_nt'])+100]\n",
    "    ref_seq_end = ref_genome_seq_ref_dict[df.loc[index]['bNumber']][int(df.loc[index]['Domain_end_nt'])-100:int(df.loc[index]['Domain_end_nt'])]\n",
    "    df.set_value(index, 'Domain_beg_in_patric', ref_genome_seq_patric_dict[df.loc[index]['patricNumber']].find(ref_seq_beg))\n",
    "    df.set_value(index, 'Domain_end_in_patric', ref_genome_seq_patric_dict[df.loc[index]['patricNumber']].find(ref_seq_end) + 100)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "###Inside of domains\n",
    "domain_dict = {}\n",
    "for index in df.index:\n",
    "    if df.loc[index]['Domain_beg_in_patric'] > 100:\n",
    "        try:\n",
    "            domain_dict[df.loc[index]['patricNumber']].extend(list(range(df.loc[index]['Domain_beg_in_patric'], df.loc[index]['Domain_end_in_patric'])))\n",
    "        except:\n",
    "            domain_dict[df.loc[index]['patricNumber']] = list(range(df.loc[index]['Domain_beg_in_patric'], df.loc[index]['Domain_end_in_patric']))\n",
    "    else:\n",
    "        if df.loc[index]['Domain_end_in_patric'] < 100:\n",
    "            pass\n",
    "        else:\n",
    "            try:\n",
    "                domain_dict[df.loc[index]['patricNumber']].extend(list(range(100, df.loc[index]['Domain_end_in_patric'])))\n",
    "            except:\n",
    "                domain_dict[df.loc[index]['patricNumber']] = list(range(100, df.loc[index]['Domain_end_in_patric']))\n",
    "\n",
    "to_delete = []        \n",
    "for gene_name in domain_dict.keys():\n",
    "    if gene_name not in dicty_to_analyze.keys():\n",
    "        to_delete.append(gene_name)\n",
    "for gene_name in to_delete:\n",
    "    del domain_dict[gene_name]\n",
    "print('Inside domains: {}'.format(len(domain_dict.keys())))\n",
    "\n",
    "\n",
    "###Post domain dicts\n",
    "post_domain_dict = {}\n",
    "for index in df.index:\n",
    "    try:\n",
    "        post_domain_dict[df.loc[index]['patricNumber']].append(df.loc[index]['Domain_end_in_patric'])\n",
    "    except:\n",
    "        post_domain_dict[df.loc[index]['patricNumber']] = [df.loc[index]['Domain_end_in_patric']]\n",
    "        \n",
    "to_delete = []        \n",
    "for gene_name in post_domain_dict.keys():\n",
    "    if gene_name not in dicty_to_analyze.keys():\n",
    "        to_delete.append(gene_name)\n",
    "for gene_name in to_delete:\n",
    "    del post_domain_dict[gene_name]\n",
    "\n",
    "for i,j in post_domain_dict.items():\n",
    "    tempy = []\n",
    "    for loc in j:\n",
    "        if loc+150 > len(seq_dicty[base_genome][i])-50:\n",
    "            if loc+30 > len(seq_dicty[base_genome][i]):\n",
    "                pass\n",
    "            else:\n",
    "                tempy.extend(list(range(loc+30, len(seq_dicty[base_genome][i])-50))) #####IMPORTANT\n",
    "        else:\n",
    "            tempy.extend(list(range(loc+30, loc+150))) #####IMPORTANT\n",
    "    post_domain_dict[i] = tempy\n",
    "print('Post domains: {}'.format(len(post_domain_dict.keys())))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Copy and pasted code above, just making a few changes to initial parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dicty_to_analyze = dicty_hyphy #####If this changes, make sure to manually change file_name_modifiers\n",
    "threshold=-4.5\n",
    "upper_threshold=-20\n",
    "# subset_names=domain_dict.keys()#####CRITICAL\n",
    "subset_names=post_domain_dict.keys()#####CRITICAL\n",
    "only_conserved_aa=True\n",
    "# position_limitation_dict=domain_dict####CRITICAL\n",
    "position_limitation_dict=post_domain_dict####CRITICAL\n",
    "controls=('CODON', 'TRINT')\n",
    "nt_strength_restriction='3'\n",
    "\n",
    "#########################################################\n",
    "####Putative aSD sequence to use in considering hybridization energies\n",
    "with open('../Data/json-energyRef-CCUCCU.txt') as infile:\n",
    "    fragment_energy_dict = json.load(infile)\n",
    "asd_seq = 'CCUCCU'\n",
    "\n",
    "\n",
    "actual_ratios_codon = []\n",
    "shuffled_ratios_codon = []\n",
    "actual_ratios_trint = []\n",
    "shuffled_ratios_trint = []\n",
    "for i in range(20):\n",
    "    print('#####{}'.format(i))\n",
    "    sds, others, analyzed_sd_dict, analyzed_ctrl_dict = paired_nt_analysis(dicty_to_analyze,\\\n",
    "                                     seq_dicty[base_genome],\\\n",
    "                                     fragment_energy_dict,\\\n",
    "                                     asd_seq,\\\n",
    "                                     conservation_dicty,\\\n",
    "                                     nt_conservation_dicty,\\\n",
    "                                     threshold=threshold,\\\n",
    "                                     upper_threshold=upper_threshold,\\\n",
    "                                     randomization=False,\\\n",
    "                                     subset_names=subset_names,\\\n",
    "                                     only_conserved_aa=only_conserved_aa,\\\n",
    "                                     position_limitation_dict=position_limitation_dict,\\\n",
    "                                     internal_control=controls[0],\\\n",
    "                                     nt_strength_restriction=nt_strength_restriction)\n",
    "    actual_ratios_codon.append(np.mean(sds)/np.mean(others))\n",
    "    print('Lengths:{}, {}, Means:{}, {}, Ratio:{}'.format(len(sds), len(others), np.mean(sds), np.mean(others), np.mean(sds)/np.mean(others)))\n",
    "    sds, others, analyzed_sd_dict, analyzed_ctrl_dict = paired_nt_analysis(dicty_to_analyze,\\\n",
    "                                     seq_dicty[base_genome],\\\n",
    "                                     fragment_energy_dict,\\\n",
    "                                     asd_seq,\\\n",
    "                                     conservation_dicty,\\\n",
    "                                     nt_conservation_dicty,\\\n",
    "                                     threshold=threshold,\\\n",
    "                                     upper_threshold=upper_threshold,\\\n",
    "                                     randomization=True,\\\n",
    "                                     subset_names=subset_names,\\\n",
    "                                     only_conserved_aa=only_conserved_aa,\\\n",
    "                                     position_limitation_dict=position_limitation_dict,\\\n",
    "                                     internal_control=controls[0],\\\n",
    "                                     nt_strength_restriction=nt_strength_restriction)\n",
    "    shuffled_ratios_codon.append(np.mean(sds)/np.mean(others))\n",
    "    \n",
    "    sds, others, analyzed_sd_dict, analyzed_ctrl_dict = paired_nt_analysis(dicty_to_analyze,\\\n",
    "                                     seq_dicty[base_genome],\\\n",
    "                                     fragment_energy_dict,\\\n",
    "                                     asd_seq,\\\n",
    "                                     conservation_dicty,\\\n",
    "                                     nt_conservation_dicty,\\\n",
    "                                     threshold=threshold,\\\n",
    "                                     upper_threshold=upper_threshold,\\\n",
    "                                     randomization=False,\\\n",
    "                                     subset_names=subset_names,\\\n",
    "                                     only_conserved_aa=only_conserved_aa,\\\n",
    "                                     position_limitation_dict=position_limitation_dict,\\\n",
    "                                     internal_control=controls[1],\\\n",
    "                                     nt_strength_restriction=nt_strength_restriction)\n",
    "    actual_ratios_trint.append(np.mean(sds)/np.mean(others))\n",
    "    print('Lengths:{}, {}, Means:{}, {}, Ratio:{}'.format(len(sds), len(others), np.mean(sds), np.mean(others), np.mean(sds)/np.mean(others)))\n",
    "    sds, others, analyzed_sd_dict, analyzed_ctrl_dict = paired_nt_analysis(dicty_to_analyze,\\\n",
    "                                     seq_dicty[base_genome],\\\n",
    "                                     fragment_energy_dict,\\\n",
    "                                     asd_seq,\\\n",
    "                                     conservation_dicty,\\\n",
    "                                     nt_conservation_dicty,\\\n",
    "                                     threshold=threshold,\\\n",
    "                                     upper_threshold=upper_threshold,\\\n",
    "                                     randomization=True,\\\n",
    "                                     subset_names=subset_names,\\\n",
    "                                     only_conserved_aa=only_conserved_aa,\\\n",
    "                                     position_limitation_dict=position_limitation_dict,\\\n",
    "                                     internal_control=controls[1],\\\n",
    "                                     nt_strength_restriction=nt_strength_restriction)\n",
    "    shuffled_ratios_trint.append(np.mean(sds)/np.mean(others))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if position_limitation_dict == domain_dict:\n",
    "    dfA = pd.DataFrame({'Actual_rats_codon_domain':actual_ratios_codon})\n",
    "    dfB = pd.DataFrame({'Actual_rats_trint_domain':actual_ratios_trint})\n",
    "elif position_limitation_dict == post_domain_dict:\n",
    "    dfC = pd.DataFrame({'Actual_rats_codon_post_domain':actual_ratios_codon})\n",
    "    dfD = pd.DataFrame({'Actual_rats_trint_post_domain':actual_ratios_trint})\n",
    "else:\n",
    "    print('Go back and change something!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df = pd.concat([dfA[['Actual_rats_codon_domain']],dfB[['Actual_rats_trint_domain']],\\\n",
    "               dfC[['Actual_rats_codon_post_domain']],dfD[['Actual_rats_trint_post_domain']]], axis=1)\n",
    "df.to_csv('../Results/temp_df_domain_analysis_cotrans_NTrestriction={}_ratios.csv'.format(nt_strength_restriction), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing gene expression differences systematically"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### First, tediously get a dictionary linking names in PATRIC annotations to names in RefSeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "genome_ref = list(SeqIO.parse('/Users/adamhockenberry/Projects/2017/Genome_database/Data/GBFs/511145.12.RefSeq.gbf', 'genbank'))[0]\n",
    "genome_pat = list(SeqIO.parse('/Users/adamhockenberry/Projects/2017/Genome_database/Data/GBFs/511145.12.PATRIC.gbf', 'genbank'))[0]\n",
    "\n",
    "locs_dict_ref_positive = {}\n",
    "locs_dict_ref_negative = {}\n",
    "\n",
    "locs_dict_pat_positive = {}\n",
    "locs_dict_pat_negative = {}\n",
    "\n",
    "ref_genome_seq_ref_dict = {}\n",
    "ref_genome_seq_patric_dict = {}\n",
    "\n",
    "for feature_pat in genome_pat.features[:]:\n",
    "    if feature_pat.type == 'CDS':\n",
    "        name_pat = feature_pat.qualifiers['locus_tag'][0]\n",
    "        strand_pat = feature_pat.strand\n",
    "        if strand_pat == 1:\n",
    "            locs_dict_pat_positive[name_pat] = list(range(feature_pat.location.start, feature_pat.location.end))\n",
    "            ref_genome_seq_patric_dict[name_pat] = genome_pat.seq[feature_pat.location.start:feature_pat.location.end]\n",
    "        elif strand_pat == -1:\n",
    "            locs_dict_pat_negative[name_pat] = list(range(feature_pat.location.start, feature_pat.location.end))\n",
    "            ref_genome_seq_patric_dict[name_pat] = genome_pat.seq[feature_pat.location.start:feature_pat.location.end].reverse_complement()\n",
    "\n",
    "for feature_ref in genome_ref.features[:]:\n",
    "    if feature_ref.type == 'CDS':\n",
    "        name_ref = feature_ref.qualifiers['locus_tag'][0]\n",
    "        strand_ref = feature_ref.strand\n",
    "        if strand_ref == 1:\n",
    "            locs_dict_ref_positive[name_ref] = list(range(feature_ref.location.start, feature_ref.location.end))\n",
    "            ref_genome_seq_ref_dict[name_ref] = genome_ref.seq[feature_ref.location.start:feature_ref.location.end]\n",
    "\n",
    "        elif strand_ref == -1:\n",
    "            locs_dict_ref_negative[name_ref] = list(range(feature_ref.location.start, feature_ref.location.end))\n",
    "            ref_genome_seq_ref_dict[name_ref] = genome_ref.seq[feature_ref.location.start:feature_ref.location.end].reverse_complement()\n",
    "            \n",
    "print(len(locs_dict_pat_positive.keys()))\n",
    "print(len(locs_dict_pat_negative.keys()))\n",
    "print(len(locs_dict_ref_positive.keys()))\n",
    "print(len(locs_dict_ref_negative.keys()))   \n",
    "\n",
    "reference_name_dict = {}\n",
    "for name_pat, loc_pat in sorted(locs_dict_pat_positive.items()):\n",
    "    for name_ref, loc_ref in sorted(locs_dict_ref_positive.items()):\n",
    "        if len(set(loc_pat)&set(loc_ref)) / len(loc_pat) > 0.9:\n",
    "            reference_name_dict[name_ref] = name_pat\n",
    "            del locs_dict_ref_positive[name_ref]\n",
    "            break\n",
    "            \n",
    "for name_pat, loc_pat in sorted(locs_dict_pat_negative.items()):\n",
    "    for name_ref, loc_ref in sorted(locs_dict_ref_negative.items()):\n",
    "        if len(set(loc_pat)&set(loc_ref)) / len(loc_pat) > 0.9:\n",
    "            reference_name_dict[name_ref] = name_pat\n",
    "            del locs_dict_ref_negative[name_ref]\n",
    "            break\n",
    "print(len(reference_name_dict.keys()), len(ref_genome_seq_patric_dict.keys()), len(ref_genome_seq_ref_dict.keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load in dataset of protein abundance from PaxDB (integrated dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv('/Users/adamhockenberry/Downloads/511145/511145-WHOLE_ORGANISM-integrated.txt',\\\n",
    "                 index_col = 'string_external_id', sep='\\t', skiprows=10)\n",
    "expression_dict = {}\n",
    "for index in df.index:\n",
    "    temp_name = index.split('.')[-1]\n",
    "    try:\n",
    "        expression_dict[reference_name_dict[temp_name]] = df.loc[index]['abundance']\n",
    "    except KeyError:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get list of gene expression values in my dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dict_to_analyze = dicty_hyphy\n",
    "\n",
    "full_listy = []\n",
    "for key in expression_dict.keys():\n",
    "    if key in dict_to_analyze.keys():\n",
    "        full_listy.append(expression_dict[key])\n",
    "print(len(full_listy))\n",
    "full_listy_to_plot = sorted(full_listy)[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "percentile_20 = np.percentile(full_listy, 20)\n",
    "percentile_40 = np.percentile(full_listy, 40)\n",
    "percentile_60 = np.percentile(full_listy, 60)\n",
    "percentile_80 = np.percentile(full_listy, 80)\n",
    "\n",
    "expression_quantile_dict = {}\n",
    "expression_quantile_dict['lowest'] = []\n",
    "expression_quantile_dict['low'] = []\n",
    "expression_quantile_dict['mid'] = []\n",
    "expression_quantile_dict['high'] = []\n",
    "expression_quantile_dict['highest'] = []\n",
    "\n",
    "for seq_name in dict_to_analyze:\n",
    "    if seq_name not in expression_dict.keys():\n",
    "        continue\n",
    "    if expression_dict[seq_name] < percentile_20:\n",
    "        expression_quantile_dict['lowest'].append(seq_name)\n",
    "    elif expression_dict[seq_name] >= percentile_20 and expression_dict[seq_name] < percentile_40:\n",
    "        expression_quantile_dict['low'].append(seq_name)\n",
    "    elif expression_dict[seq_name] >= percentile_40 and expression_dict[seq_name] < percentile_60:\n",
    "        expression_quantile_dict['mid'].append(seq_name)\n",
    "    elif expression_dict[seq_name] >= percentile_60 and expression_dict[seq_name] < percentile_80:\n",
    "        expression_quantile_dict['high'].append(seq_name)\n",
    "    elif expression_dict[seq_name] >= percentile_80 :\n",
    "        expression_quantile_dict['highest'].append(seq_name)\n",
    "    \n",
    "for key, vals in expression_quantile_dict.items():\n",
    "    print(key, len(vals))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualize some differences between the categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(5,2))\n",
    "ax.hist(full_listy_to_plot, 10000, cumulative=True, histtype='step')\n",
    "ax.set_xscale('log')\n",
    "# ax.set_yscale('log')\n",
    "ax.axvspan(percentile_20, percentile_40, color='k', alpha=0.2)\n",
    "ax.axvspan(percentile_60, percentile_80, color='k', alpha=0.2)\n",
    "ax.set_xlabel('Protein abundance')\n",
    "ax.set_ylabel('Cumulative\\ncounts')\n",
    "plt.tight_layout()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "asd_seq = 'CCUCCU'\n",
    "threshold = -4.5\n",
    "upper_threshold = -20\n",
    "\n",
    "density_dict = {}\n",
    "eligible_density_dict = {}\n",
    "exp_energy_dict = {}\n",
    "basic_conservation_dict = {}\n",
    "for category in ['lowest', 'low', 'mid', 'high', 'highest'][:]:\n",
    "    lengths = []\n",
    "    sd_nts = []\n",
    "    energies_temp = []\n",
    "    cons = []\n",
    "    eligible_sites = []\n",
    "    \n",
    "    for gene_name in expression_quantile_dict[category]:\n",
    "        if gene_name in seq_dicty[base_genome].keys():\n",
    "            seq = seq_dicty[base_genome][gene_name]\n",
    "            sd_positions_all_nts, sd_positions_start, energy_list, codon_list =\\\n",
    "                get_sd_sites(seq, asd_seq, threshold, fragment_energy_dict,\\\n",
    "                             position_limitation_list=False, upper_threshold=upper_threshold)\n",
    "            sd_nts.append(len(sd_positions_all_nts))\n",
    "            lengths.append(len(seq)-150)\n",
    "            energies_temp.extend([i for i in energy_list if i < threshold])\n",
    "            cons.append(conservation_dicty[gene_name].count('C')/len(conservation_dicty[gene_name]))\n",
    "            \n",
    "            eligible_sites_ind = []\n",
    "            for i in sd_positions_start:\n",
    "                modifiers = []\n",
    "                if i%3 == 0:\n",
    "                    modifiers = [2]\n",
    "                elif i%3 == 2:\n",
    "                    modifiers = [3]\n",
    "                elif i%3 == 1:\n",
    "                    modifiers = [1,4]\n",
    "                for modifier in modifiers:\n",
    "                    if conservation_dicty[gene_name][int(np.floor((i+modifier)/3.))] != 'C':\n",
    "                        continue\n",
    "                    codon = seq[i+modifier-2:i+modifier+1]\n",
    "                    if codon in redundant_codons:\n",
    "                        eligible_sites_ind.append(i+modifier)\n",
    "                        \n",
    "            eligible_sites_ind = list(set(eligible_sites_ind))\n",
    "            eligible_sites.append(len(eligible_sites_ind))\n",
    "            \n",
    "            \n",
    "            \n",
    "    density_dict[category] = np.array(sd_nts)/np.array(lengths)\n",
    "    exp_energy_dict[category] = energies_temp\n",
    "    basic_conservation_dict[category] = cons\n",
    "    eligible_density_dict[category] = np.array(eligible_sites)/np.array(lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3.5,3.5))\n",
    "ax = fig.add_subplot(311)\n",
    "boxes = ax.bar([0, 1, 2, 3, 4], [np.mean(density_dict[i]) for i in ['lowest', 'low', 'mid', 'high', 'highest']], facecolor='0.3')\n",
    "# ax.set_ylim(-0.01, 0.18)\n",
    "\n",
    "ax.set_xticklabels('')\n",
    "ax.set_ylabel('SD-like nt\\nfraction')\n",
    "\n",
    "for line in ax.xaxis.get_ticklines():\n",
    "    line.set_markersize(0)\n",
    "    line.set_markeredgewidth(0)\n",
    "\n",
    "ax = fig.add_subplot(312)\n",
    "boxes = ax.bar([0, 1, 2, 3, 4], [np.mean(basic_conservation_dict[i]) for i in ['lowest', 'low', 'mid', 'high', 'highest']], facecolor='0.3')\n",
    "# ax.set_ylim(-0.01, 0.18)\n",
    "ax.set_xticklabels('')\n",
    "ax.set_ylabel('Conserved AA\\nfraction')\n",
    "for line in ax.xaxis.get_ticklines():\n",
    "    line.set_markersize(0)\n",
    "    line.set_markeredgewidth(0)\n",
    "\n",
    "ax = fig.add_subplot(313)\n",
    "boxes = ax.bar([0, 1, 2, 3, 4], [np.mean(eligible_density_dict[i]) for i in ['lowest', 'low', 'mid', 'high', 'highest']], facecolor='0.3')\n",
    "ax.set_xticklabels('')\n",
    "ax.set_ylabel('Eligible SD-like\\nnt fraction')\n",
    "ax.set_xticklabels(['', 'Lowest 20%', '', '', '', 'Highest 20%'], ha='center')\n",
    "ax.set_xlabel('Gene expression bins')\n",
    "plt.tight_layout(h_pad=-0.1)\n",
    "plt.savefig('../Results/Figures/expression_summary.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#####Parameters!\n",
    "dicty_to_analyze = dicty_hyphy\n",
    "threshold=-4.5\n",
    "upper_threshold=-20\n",
    "only_conserved_aa=True\n",
    "position_limitation_dict=False\n",
    "controls=('CODON', 'TRINT')\n",
    "nt_strength_restriction='None'\n",
    "\n",
    "meta_codon = []\n",
    "meta_trint = []\n",
    "for cat in ['lowest', 'low', 'mid', 'high', 'highest']:\n",
    "    file_name_modifiers = '{}_{}_Threshold={}to{}_Conserved={}_NtRestriction={}_Subset={}'.format('hyphy', species_name, upper_threshold, threshold, only_conserved_aa, nt_strength_restriction, cat)\n",
    "    print('#####{}'.format(cat))\n",
    "    subset_names = expression_quantile_dict[cat]\n",
    "    actual_ratios_codon = []\n",
    "    shuffled_ratios_codon = {}\n",
    "    actual_ratios_trint = []\n",
    "    shuffled_ratios_trint = {}\n",
    "    for i in range(20):\n",
    "        shuffled_ratios_codon[i] = []\n",
    "        shuffled_ratios_trint[i] = []\n",
    "        print('#####{}'.format(i))\n",
    "        sds_codon, others_codon, analyzed_sd_dict, analyzed_ctrl_dict, all_sds_codon, all_others_codon = paired_nt_analysis(dicty_to_analyze,\\\n",
    "                                         seq_dicty[base_genome],\\\n",
    "                                         fragment_energy_dict,\\\n",
    "                                         asd_seq,\\\n",
    "                                         conservation_dicty,\\\n",
    "                                         nt_conservation_dicty,\\\n",
    "                                         threshold=threshold,\\\n",
    "                                         upper_threshold=upper_threshold,\\\n",
    "                                         subset_names=subset_names,\\\n",
    "                                         only_conserved_aa=only_conserved_aa,\\\n",
    "                                         position_limitation_dict=position_limitation_dict,\\\n",
    "                                         internal_control=controls[0],\\\n",
    "                                         nt_strength_restriction=nt_strength_restriction)\n",
    "        actual_ratios_codon.append(np.mean(sds_codon)/np.mean(others_codon))\n",
    "        print('Lengths:{}, {}, Means:{}, {}, Ratio:{}'.format(len(sds_codon), len(others_codon),\\\n",
    "                                                              np.mean(sds_codon), np.mean(others_codon),\\\n",
    "                                                              np.mean(sds_codon)/np.mean(others_codon)))\n",
    "        diffs = np.array(sds_codon)-np.array(others_codon)\n",
    "        print('Coin flip expt:', len([i for i in diffs if i > 0])/len([i for i in diffs if i != 0]))\n",
    "        for j in range(stat_randomizations):\n",
    "            combined = sds_codon + others_codon\n",
    "            random.shuffle(combined)\n",
    "            random_sds = combined[:len(sds_codon)]\n",
    "            random_others = combined[len(sds_codon):]\n",
    "            shuffled_ratios_codon[i].append(np.mean(random_sds)/np.mean(random_others))\n",
    "\n",
    "\n",
    "        sds_trint, others_trint, analyzed_sd_dict, analyzed_ctrl_dict, all_sds_trint, all_others_trint = paired_nt_analysis(dicty_to_analyze,\\\n",
    "                                         seq_dicty[base_genome],\\\n",
    "                                         fragment_energy_dict,\\\n",
    "                                         asd_seq,\\\n",
    "                                         conservation_dicty,\\\n",
    "                                         nt_conservation_dicty,\\\n",
    "                                         threshold=threshold,\\\n",
    "                                         upper_threshold=upper_threshold,\\\n",
    "                                         subset_names=subset_names,\\\n",
    "                                         only_conserved_aa=only_conserved_aa,\\\n",
    "                                         position_limitation_dict=position_limitation_dict,\\\n",
    "                                         internal_control=controls[1],\\\n",
    "                                         nt_strength_restriction=nt_strength_restriction)\n",
    "        actual_ratios_trint.append(np.mean(sds_trint)/np.mean(others_trint))\n",
    "        print('Lengths:{}, {}, Means:{}, {}, Ratio:{}'.format(len(sds_trint), len(others_trint),\\\n",
    "                                                              np.mean(sds_trint), np.mean(others_trint),\\\n",
    "                                                              np.mean(sds_trint)/np.mean(others_trint)))\n",
    "        diffs = np.array(sds_trint)-np.array(others_trint)\n",
    "        print('Coin flip expt:', len([i for i in diffs if i > 0])/len([i for i in diffs if i != 0]))\n",
    "        for j in range(stat_randomizations):\n",
    "            combined = sds_trint + others_trint\n",
    "            random.shuffle(combined)\n",
    "            random_sds = combined[:len(sds_trint)]\n",
    "            random_others = combined[len(sds_trint):]\n",
    "            shuffled_ratios_trint[i].append(np.mean(random_sds)/np.mean(random_others))\n",
    "    \n",
    "    \n",
    "    dfA = pd.DataFrame({'Actual_rats_codon':actual_ratios_codon})\n",
    "    dfB = pd.DataFrame({'Shuffled_rats_codon':[j[0] for i,j in shuffled_ratios_codon.items()]})\n",
    "    dfC = pd.DataFrame({'Actual_rats_trint':actual_ratios_trint})\n",
    "    dfD = pd.DataFrame({'Shuffled_rats_trint':[j[0] for i,j in shuffled_ratios_trint.items()]})\n",
    "    df = pd.concat([dfA[['Actual_rats_codon']],dfB[['Shuffled_rats_codon']],\\\n",
    "                   dfC[['Actual_rats_trint']],dfD[['Shuffled_rats_trint']]], axis=1)\n",
    "    df['N_(codon)'] = len(sds_codon)\n",
    "    df['N_(trint)'] = len(sds_trint)\n",
    "    df.to_csv('../Results/Dataframes/df_{}_ratios.csv'.format(file_name_modifiers), index=False)\n",
    "\n",
    "\n",
    "    dfA = pd.DataFrame({'SD-like(codon)':all_sds_codon})\n",
    "    dfB = pd.DataFrame({'SD-like(trint)':all_sds_trint})\n",
    "    dfC = pd.DataFrame({'Control\\n(codon)':all_others_codon})\n",
    "    dfD = pd.DataFrame({'Control\\n(context)':all_others_trint})\n",
    "    df = pd.concat([dfA[['SD-like(codon)']],dfB[['SD-like(trint)']], dfC[['Control\\n(codon)']], dfD[['Control\\n(context)']]], axis=1)\n",
    "    df.to_csv('../Results/Dataframes/df_{}_all.csv'.format(file_name_modifiers), index=False)\n",
    "\n",
    "    dist_of_codon_means = []\n",
    "    for j in range(stat_randomizations):\n",
    "        temp = []\n",
    "        for i in range(mean_randomizations):\n",
    "            temp.append(shuffled_ratios_codon[i][j])\n",
    "        dist_of_codon_means.append(np.mean(temp))\n",
    "\n",
    "    dist_of_trint_means = []\n",
    "    for j in range(stat_randomizations):\n",
    "        temp = []\n",
    "        for i in range(mean_randomizations):\n",
    "            temp.append(shuffled_ratios_trint[i][j])\n",
    "        dist_of_trint_means.append(np.mean(temp))\n",
    "\n",
    "    dfA = pd.DataFrame({'Names':['Actual']+list(range(len(dist_of_codon_means)))})\n",
    "    dfB = pd.DataFrame({'Values(codon)':[np.mean(actual_ratios_codon)]+dist_of_codon_means})\n",
    "    dfC = pd.DataFrame({'Values(trint)':[np.mean(actual_ratios_trint)]+dist_of_trint_means})\n",
    "    df = pd.concat([dfA[['Names']],dfB[['Values(codon)']], dfC[['Values(trint)']]], axis=1)\n",
    "    df.to_csv('../Results/Dataframes/df_{}_stat_sig.csv'.format(file_name_modifiers), index=False)\n",
    "\n",
    "    \n",
    "    \n",
    "    meta_codon.append(actual_ratios_codon)\n",
    "    meta_trint.append(actual_ratios_trint)\n",
    "    print(len(sds_codon), np.mean(actual_ratios_codon), np.mean(actual_ratios_trint))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.violinplot(meta_codon, showextrema=False)\n",
    "\n",
    "\n",
    "plt.figure()\n",
    "plt.violinplot(meta_trint, showextrema=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# DEPRECATED CODE THAT I'D LIKE TO HANG ON TO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dicty_to_analyze = dicty_hyphy\n",
    "threshold=-4.5\n",
    "upper_threshold=-20\n",
    "subset_names=False\n",
    "only_conserved_aa=True\n",
    "three_prime_analysis=False\n",
    "controls=('CODON', 'TRINT')\n",
    "# nt_strength_restriction=''\n",
    "nt_strength_restriction='CURRENT_IS_WEAKEST'\n",
    "\n",
    "#########################################################\n",
    "####Putative aSD sequence to use in considering hybridization energies\n",
    "with open('../Data/json-energyRef-CCUCCU.txt') as infile:\n",
    "    fragment_energy_dict = json.load(infile)\n",
    "asd_seq = 'CCUCCU'\n",
    "\n",
    "\n",
    "x_sds_all = []\n",
    "x_controls_all=[]\n",
    "for i in range(2):\n",
    "    sds, others, analyzed_sd_dict, analyzed_ctrl_dict = paired_nt_analysis(dicty_to_analyze,\\\n",
    "                                     seq_dicty[base_genome],\\\n",
    "                                     fragment_energy_dict,\\\n",
    "                                     asd_seq,\\\n",
    "                                     conservation_dicty,\\\n",
    "                                     threshold=threshold,\\\n",
    "                                     upper_threshold=upper_threshold,\\\n",
    "                                     randomization=False,\\\n",
    "                                     subset_names=subset_names,\\\n",
    "                                     only_conserved_aa=only_conserved_aa,\\\n",
    "                                     three_prime_analysis=three_prime_analysis,\\\n",
    "                                     internal_control=controls[0],\\\n",
    "                                     nt_strength_restriction=nt_strength_restriction)\n",
    "    print(len(sds), len(others))\n",
    "    nts_sds = []\n",
    "    nts_controls = []\n",
    "    for gene_name, positions in analyzed_sd_dict.items():\n",
    "        for i in positions:\n",
    "            ec_nt = seq_dicty[base_genome][gene_name][i]\n",
    "            matches = []\n",
    "            for genome in ordered_names:\n",
    "                if seq_dicty[genome][gene_name][i] == ec_nt:\n",
    "                    matches.append(1)\n",
    "                else:\n",
    "                    matches.append(0)\n",
    "            nts_sds.append(matches)\n",
    "\n",
    "    for gene_name, positions in analyzed_ctrl_dict.items():\n",
    "        for i in positions:\n",
    "            ec_nt = seq_dicty[base_genome][gene_name][i]\n",
    "            matches = []\n",
    "            for genome in ordered_names:\n",
    "                if seq_dicty[genome][gene_name][i] == ec_nt:\n",
    "                    matches.append(1)\n",
    "                else:\n",
    "                    matches.append(0)\n",
    "            nts_controls.append(matches)\n",
    "    nts_sds = nts_sds * np.exp(np.linspace(10, 1, 61))\n",
    "    nts_sds = sorted(nts_sds, key=lambda x: np.sum(x))[::-1]\n",
    "    nts_sds = nts_sds / np.exp(np.linspace(10, 1, 61))\n",
    "    nts_controls = nts_controls * np.exp(np.linspace(10, 1, 61))\n",
    "    nts_controls = sorted(nts_controls, key=lambda x: np.sum(x))[::-1]\n",
    "    nts_controls = nts_controls / np.exp(np.linspace(10, 1, 61))\n",
    "    \n",
    "    x_sds_all.append(np.sum(nts_sds, axis=0)/np.shape(nts_sds)[0])\n",
    "    x_controls_all.append(np.sum(nts_controls, axis=0)/np.shape(nts_sds)[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Making figures from the above data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "tree = Phylo.read('../Data/Tree_files/Order_Enterobacterales_Mixed/RAxML_bestTree.MixedSet108seqs', 'newick')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "####Order strictly via branch length distance to base_genome\n",
    "ordered_names = []\n",
    "for i in tree.get_terminals():\n",
    "    ordered_names.append((i.name, tree.distance(i.name, base_genome)))\n",
    "ordered_names = sorted(ordered_names, key=lambda x: x[1])\n",
    "ordered_names = [i[0] for i in ordered_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###This needs to be written recursively to scale! in any meaningful way\n",
    "###First order by the closest common ancestor to base_genome\n",
    "initial_ordering = []\n",
    "for i in tree.get_terminals():\n",
    "    initial_ordering.append((i.name, len(tree.trace(tree.common_ancestor(i.name, base_genome), base_genome))))\n",
    "initial_ordering = sorted(initial_ordering, key=lambda x: x[1])\n",
    "###Resolve all ties by making whoever is closest to base_genome the top OR by making that call based on \n",
    "###who is closest via clades rather than distance. hmm\n",
    "###Strictly speaking distance makes sense but can produce results that are not consistent with rotation\n",
    "ordered_names = []\n",
    "for i in range(1, max([dist for term,dist in initial_ordering])+1):\n",
    "    tempy = [(term, dist) for term,dist in initial_ordering if dist == i]\n",
    "    if len(tempy) == 1:\n",
    "        ordered_names.append((tempy[0]))\n",
    "    else:\n",
    "        temp_ordering = []\n",
    "        for blah in tempy:\n",
    "            temp_ordering.append((blah[0],len(tree.trace(blah[0], base_genome))))\n",
    "        temp_ordering = sorted(temp_ordering, key=lambda x: x[1])\n",
    "        ordered_names.extend(temp_ordering)\n",
    "ordered_names = [i for i,j in ordered_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(12, 5)) \n",
    "gs = gridspec.GridSpec(1, 2, width_ratios=[4, 1]) \n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax1.matshow(np.array(nts_sds).T, cmap='Greys_r', aspect='auto', )\n",
    "ax1.set_yticklabels('')\n",
    "ax1.set_xticklabels('')\n",
    "for line in ax1.xaxis.get_ticklines():\n",
    "    line.set_markersize(0)\n",
    "    line.set_markeredgewidth(0)\n",
    "for line in ax1.yaxis.get_ticklines():\n",
    "    line.set_markersize(0)\n",
    "    line.set_markeredgewidth(0)\n",
    "ax1.grid(False)\n",
    "ax1.set_ylabel('Species\\n(ordered phylogenetically)')\n",
    "ax1.set_xlabel('Sample of 4-fold redundant sites')\n",
    "ax1.spines['right'].set_visible(True)\n",
    "ax1.spines['top'].set_visible(True)\n",
    "ax1.spines['left'].set_visible(True)\n",
    "ax1.spines['bottom'].set_visible(True)\n",
    "\n",
    "ax2 = plt.subplot(gs[1], sharey=ax1)\n",
    "ax2.plot(np.sum(nts_sds, axis=0)/np.shape(nts_sds)[0], list(range(len(ordered_names))), c='k')\n",
    "ax2.set_xlabel('ANI across 4-fold\\nredundant sites')\n",
    "# ax2.set_xlim(0.25, 1.01)\n",
    "ax2.set_xticks([0.5, 1])\n",
    "for line in ax2.yaxis.get_ticklines():\n",
    "    line.set_markersize(0)\n",
    "    line.set_markeredgewidth(0)\n",
    "ax2.set_ylim(61,0)\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Results/Figures/ANI_example.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(6,6))\n",
    "yvals = list(range(len(ordered_names)))[::-1]\n",
    "# x_sds = np.sum(nts_sds, axis=0)/np.shape(nts_sds)[0]\n",
    "# x_controls = np.sum(nts_controls, axis=0)/np.shape(nts_sds)[0]\n",
    "x_sds = np.mean(np.array(x_sds_all), axis=0)\n",
    "x_controls = np.mean(np.array(x_controls_all), axis=0)\n",
    "ax.plot(x_sds, yvals, label='SD-like', c='steelblue')\n",
    "ax.plot(x_controls, yvals, label='Codon control', c='gray')\n",
    "ax.fill_betweenx(yvals, x_sds, x_controls, where=x_sds>=x_controls, facecolor='steelblue', alpha=0.25)\n",
    "ax.fill_betweenx(yvals, x_sds, x_controls, where=x_sds<x_controls, facecolor='lightgray', alpha=0.5)\n",
    "ax.set_xlim(0.2,1.01)\n",
    "ax.legend(fontsize=16, loc='best', framealpha=1)\n",
    "ax.set_xlabel('ANI across 4-fold\\nredundant sites')\n",
    "ax.set_ylabel('Species (ordered phylogenetically)')\n",
    "ax.set_yticklabels('');\n",
    "# ax.set_ylabel('')\n",
    "\n",
    "for line in ax.yaxis.get_ticklines():\n",
    "    line.set_markersize(0)\n",
    "    line.set_markeredgewidth(0)\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Results/Figures/ANI_SD_CONSERVED_STRONGEST.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(6,6))\n",
    "yvals = list(range(len(ordered_names)))[::-1]\n",
    "# x_sds = np.sum(nts_sds, axis=0)/np.shape(nts_sds)[0]\n",
    "# x_controls = np.sum(nts_controls, axis=0)/np.shape(nts_sds)[0]\n",
    "x_sds = np.mean(np.array(x_sds_all), axis=0)\n",
    "x_controls = np.mean(np.array(x_controls_all), axis=0)\n",
    "\n",
    "for i in range(len(yvals)):\n",
    "    if x_controls[i] > x_sds[i]:\n",
    "        ax.plot([x_sds[i], x_controls[i]], [yvals[i], yvals[i]], c='lightgray', linewidth='3', alpha=0.5)\n",
    "    else:\n",
    "        ax.plot([x_sds[i], x_controls[i]], [yvals[i], yvals[i]], c='steelblue', linewidth='3', alpha=0.5)\n",
    "ax.plot(x_sds, yvals, marker='o', linestyle='', label='SD-like', c='steelblue')\n",
    "ax.plot(x_controls, yvals, marker='o', linestyle='', label='Control', c='gray')\n",
    "\n",
    "# ax.fill_betweenx(yvals, x_sds, x_controls, where=x_sds>=x_controls, facecolor='steelblue', alpha=0.25)\n",
    "# ax.fill_betweenx(yvals, x_sds, x_controls, where=x_sds<x_controls, facecolor='darkorange', alpha=0.25)\n",
    "ax.set_xlim(0.2,1.01)\n",
    "# ax.legend(fontsize=20, loc='best', framealpha=1)\n",
    "ax.set_xlabel('ANI across 4-fold\\nredundant sites')\n",
    "ax.set_ylabel('Species (ordered phylogenetically)')\n",
    "ax.set_yticklabels('');\n",
    "# ax.set_ylabel('')\n",
    "\n",
    "for line in ax.yaxis.get_ticklines():\n",
    "    line.set_markersize(0)\n",
    "    line.set_markeredgewidth(0)\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Results/Figures/ANI_SD_CONSERVED_WEAKEST.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For contingency tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ec_consensus = 0\n",
    "ec_not_consensus = 0\n",
    "for i in nts_sds:\n",
    "    if np.median(i) == 1:\n",
    "        ec_consensus += 1\n",
    "    elif np.median(i) == 0:\n",
    "        ec_not_consensus += 1\n",
    "print(ec_consensus, ec_not_consensus)\n",
    "ec_consensus = 0\n",
    "ec_not_consensus = 0\n",
    "for i in nts_controls:\n",
    "    if np.median(i) == 1:\n",
    "        ec_consensus += 1\n",
    "    elif np.median(i) == 0:\n",
    "        ec_not_consensus += 1\n",
    "print(ec_consensus, ec_not_consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "228px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
